// @remotion/studio/dist/components/Timeline/TimelineSequenceName.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TimelineSequenceName = void 0;
const jsx_runtime_1 = require("react/jsx-runtime");
const react_1 = require("react");
const TimelineSelection_1 = require("./TimelineSelection");
const MAX_DISPLAY_NAME_LENGTH = 1000;
const getTruncatedDisplayName = (displayName) => {
    if (displayName.length > MAX_DISPLAY_NAME_LENGTH) {
        return displayName.slice(0, MAX_DISPLAY_NAME_LENGTH) + '...';
    }
    return displayName;
};
const TimelineSequenceName = ({ sequence, selected, containsSelection }) => {
    const [hovered, setHovered] = (0, react_1.useState)(false);
    const { documentationLink } = sequence;
    const hoverable = !TimelineSelection_1.SELECTION_ENABLED && documentationLink !== null;
    const onClick = (0, react_1.useCallback)(() => {
        if (documentationLink === null) {
            return;
        }
        window.open(documentationLink, '_blank', 'noopener,noreferrer');
    }, [documentationLink]);
    const onPointerEnter = (0, react_1.useCallback)(() => setHovered(true), []);
    const onPointerLeave = (0, react_1.useCallback)(() => setHovered(false), []);
    const style = (0, react_1.useMemo)(() => {
        const hoverEffect = hovered && hoverable;
        return {
            alignItems: 'center',
            alignSelf: 'stretch',
            ...(0, TimelineSelection_1.getTimelineSelectedLabelStyle)(selected, false),
            display: 'inline-flex',
            fontSize: 12,
            whiteSpace: 'nowrap',
            textOverflow: 'ellipsis',
            overflow: 'hidden',
            color: (0, TimelineSelection_1.getTimelineColor)(selected, false),
            userSelect: 'none',
            WebkitUserSelect: 'none',
            textDecoration: hoverEffect ? 'underline' : 'none',
            cursor: hoverEffect ? 'pointer' : undefined,
            boxShadow: containsSelection && !selected
                ? `inset 0 0 0 2px ${TimelineSelection_1.TIMELINE_SELECTED_LABEL_BACKGROUND}`
                : undefined,
        };
    }, [hovered, hoverable, selected, containsSelection]);
    const text = getTruncatedDisplayName(sequence.displayName) || '<Sequence>';
    return (jsx_runtime_1.jsx("div", { onPointerEnter: hoverable ? onPointerEnter : undefined, onPointerLeave: hoverable ? onPointerLeave : undefined, title: documentationLink ? `Open documentation: ${documentationLink}` : text, style: style, onClick: hoverable ? onClick : undefined, children: text }));
};
exports.TimelineSequenceName = TimelineSequenceName;
