// @remotion/studio/dist/components/Timeline/TimelineMediaInfo.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.TimelineMediaInfo = exports.openTimelineAssetLink = exports.getTimelineAssetLinkInfo = void 0;
const jsx_runtime_1 = require("react/jsx-runtime");
const react_1 = __importStar(require("react"));
const remotion_1 = require("remotion");
const colors_1 = require("../../helpers/colors");
const url_state_1 = require("../../helpers/url-state");
const TimelineSelection_1 = require("./TimelineSelection");
const lineStyle = {
    whiteSpace: 'nowrap',
    overflow: 'hidden',
    textOverflow: 'ellipsis',
    minWidth: 0,
    fontSize: 12,
    color: colors_1.VERY_LIGHT_TEXT,
    display: 'inline-block',
};
const getTimelineAssetLinkInfo = (src) => {
    const staticBase = typeof window === 'undefined' ? null : window.remotion_staticBase;
    if (staticBase && src.startsWith(staticBase + '/')) {
        const assetPath = src.slice(staticBase.length + 1);
        return {
            kind: 'local',
            assetPath: decodeURIComponent(assetPath),
            title: decodeURIComponent(assetPath),
        };
    }
    if (src.startsWith('http://') ||
        src.startsWith('https://') ||
        src.startsWith('//')) {
        try {
            const url = new URL(src.startsWith('//') ? 'https:' + src : src);
            return { kind: 'remote', href: src, title: url.hostname };
        }
        catch (_a) {
            return { kind: 'remote', href: src, title: src };
        }
    }
    return null;
};
exports.getTimelineAssetLinkInfo = getTimelineAssetLinkInfo;
const openTimelineAssetLink = (linkInfo, setCanvasContent) => {
    if (linkInfo.kind === 'local') {
        setCanvasContent({ type: 'asset', asset: linkInfo.assetPath });
        (0, url_state_1.pushUrl)(`/assets/${linkInfo.assetPath}`);
        return;
    }
    window.open(linkInfo.href, '_blank', 'noopener,noreferrer');
};
exports.openTimelineAssetLink = openTimelineAssetLink;
const useAssetLink = (src) => {
    const { setCanvasContent } = react_1.default.useContext(remotion_1.Internals.CompositionSetters);
    const [hovered, setHovered] = (0, react_1.useState)(false);
    const linkInfo = (0, react_1.useMemo)(() => (0, exports.getTimelineAssetLinkInfo)(src), [src]);
    const interactive = !TimelineSelection_1.SELECTION_ENABLED && linkInfo !== null;
    const onClick = (0, react_1.useCallback)((e) => {
        if (!linkInfo) {
            return;
        }
        e.preventDefault();
        e.stopPropagation();
        (0, exports.openTimelineAssetLink)(linkInfo, setCanvasContent);
    }, [linkInfo, setCanvasContent]);
    const onPointerEnter = (0, react_1.useCallback)(() => setHovered(true), []);
    const onPointerLeave = (0, react_1.useCallback)(() => setHovered(false), []);
    const fileNameStyle = (0, react_1.useMemo)(() => ({
        ...lineStyle,
        color: interactive && hovered ? colors_1.LIGHT_TEXT : colors_1.VERY_LIGHT_TEXT,
        cursor: interactive ? 'pointer' : undefined,
        textDecoration: 'none',
        display: 'inline-block',
        overflow: 'hidden',
        whiteSpace: 'pre',
        textOverflow: 'ellipsis',
        userSelect: 'none',
        WebkitUserSelect: 'none',
    }), [interactive, hovered]);
    return {
        linkInfo,
        interactive,
        onClick,
        onPointerEnter,
        onPointerLeave,
        fileNameStyle,
    };
};
const TimelineMediaInfo = ({ src }) => {
    const fileName = (0, react_1.useMemo)(() => remotion_1.Internals.getAssetDisplayName(src), [src]);
    const { linkInfo, interactive, onClick, onPointerEnter, onPointerLeave, fileNameStyle, } = useAssetLink(src);
    return (jsx_runtime_1.jsx("div", { style: fileNameStyle, title: linkInfo ? linkInfo.title : fileName, onClick: interactive ? onClick : undefined, onPointerEnter: interactive ? onPointerEnter : undefined, onPointerLeave: interactive ? onPointerLeave : undefined, children: fileName }));
};
exports.TimelineMediaInfo = TimelineMediaInfo;
