// @remotion/studio/dist/components/Timeline/TimelineSelection.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.useTimelineRowContainsSelection = exports.useTimelineKeyframeSelection = exports.useTimelineRowSelection = exports.useTimelineSelection = exports.TimelineSelectionProvider = exports.SELECTION_ENABLED = exports.TIMELINE_SELECTED_TRACK_HIGHLIGHT_STYLE = exports.getTimelineColor = exports.getTimelineSelectedLabelStyle = exports.TIMELINE_SELECTED_LABEL_HORIZONTAL_PADDING = exports.TIMELINE_SELECTED_LABEL_TEXT = exports.TIMELINE_SELECTED_LABEL_BACKGROUND = exports.TIMELINE_SELECTED_BACKGROUND = void 0;
const jsx_runtime_1 = require("react/jsx-runtime");
const studio_shared_1 = require("@remotion/studio-shared");
const react_1 = require("react");
const client_id_1 = require("../../helpers/client-id");
const timeline_layout_1 = require("../../helpers/timeline-layout");
const timeline_node_path_key_1 = require("../../helpers/timeline-node-path-key");
exports.TIMELINE_SELECTED_BACKGROUND = '#3B3F42';
exports.TIMELINE_SELECTED_LABEL_BACKGROUND = '#B0B0B0';
exports.TIMELINE_SELECTED_LABEL_TEXT = 'black';
exports.TIMELINE_SELECTED_LABEL_HORIZONTAL_PADDING = 2;
const getTimelineSelectedLabelStyle = (selected, subcategory) => {
    return {
        paddingLeft: exports.TIMELINE_SELECTED_LABEL_HORIZONTAL_PADDING,
        paddingRight: exports.TIMELINE_SELECTED_LABEL_HORIZONTAL_PADDING,
        ...(selected
            ? {
                backgroundColor: subcategory
                    ? 'rgba(255, 255, 255, 0.1)'
                    : exports.TIMELINE_SELECTED_LABEL_BACKGROUND,
            }
            : {}),
    };
};
exports.getTimelineSelectedLabelStyle = getTimelineSelectedLabelStyle;
const getTimelineColor = (selected, subcategory) => {
    return selected && !subcategory
        ? exports.TIMELINE_SELECTED_LABEL_TEXT
        : 'rgba(255, 255, 255, 0.8)';
};
exports.getTimelineColor = getTimelineColor;
exports.TIMELINE_SELECTED_TRACK_HIGHLIGHT_STYLE = {
    backgroundColor: exports.TIMELINE_SELECTED_BACKGROUND,
    bottom: 0,
    left: -timeline_layout_1.TIMELINE_PADDING,
    pointerEvents: 'none',
    position: 'absolute',
    right: -timeline_layout_1.TIMELINE_PADDING,
    top: 0,
};
exports.SELECTION_ENABLED = false;
const TimelineSelectionContext = (0, react_1.createContext)({
    canSelect: false,
    isSelected: () => false,
    selectItem: () => undefined,
    containsSelection: () => false,
    clearSelection: () => undefined,
});
const getTimelineSelectionKey = (item) => {
    const rowKey = (0, timeline_node_path_key_1.timelineNodePathInfoToKey)(item.nodePathInfo);
    if (item.type === 'row') {
        return rowKey;
    }
    return `${rowKey}.keyframe.${item.frame}`;
};
const TimelineSelectionProvider = ({ children }) => {
    const { previewServerState } = (0, react_1.useContext)(client_id_1.StudioServerConnectionCtx);
    const canSelect = exports.SELECTION_ENABLED &&
        previewServerState.type === 'connected' &&
        !window.remotion_isReadOnlyStudio;
    const [selectedItem, setSelectedItem] = (0, react_1.useState)(null);
    (0, react_1.useEffect)(() => {
        if (!canSelect) {
            setSelectedItem(null);
        }
    }, [canSelect]);
    const isSelected = (0, react_1.useCallback)((item) => {
        return selectedItem === null
            ? false
            : getTimelineSelectionKey(selectedItem) ===
                getTimelineSelectionKey(item);
    }, [selectedItem]);
    const selectItem = (0, react_1.useCallback)((item) => {
        if (!canSelect) {
            return;
        }
        setSelectedItem(item);
    }, [canSelect]);
    const clearSelection = (0, react_1.useCallback)(() => {
        setSelectedItem(null);
    }, []);
    const containsSelection = (0, react_1.useCallback)((nodePathInfo) => {
        if (selectedItem === null) {
            return false;
        }
        const selectedNodePath = selectedItem.nodePathInfo;
        if ((0, studio_shared_1.stringifySequenceExpandedRowKey)(selectedNodePath.sequenceSubscriptionKey) !==
            (0, studio_shared_1.stringifySequenceExpandedRowKey)(nodePathInfo.sequenceSubscriptionKey)) {
            return false;
        }
        if (selectedNodePath.index !== nodePathInfo.index) {
            return false;
        }
        // Selection must be strictly deeper than this node (i.e. a descendant),
        // not the same row.
        if (selectedNodePath.auxiliaryKeys.length <=
            nodePathInfo.auxiliaryKeys.length) {
            return false;
        }
        return nodePathInfo.auxiliaryKeys.every((key, i) => selectedNodePath.auxiliaryKeys[i] === key);
    }, [selectedItem]);
    const value = (0, react_1.useMemo)(() => ({
        canSelect,
        isSelected,
        selectItem,
        containsSelection,
        clearSelection,
    }), [canSelect, isSelected, selectItem, containsSelection, clearSelection]);
    return (jsx_runtime_1.jsx(TimelineSelectionContext.Provider, { value: value, children: children }));
};
exports.TimelineSelectionProvider = TimelineSelectionProvider;
const useTimelineSelection = () => {
    return (0, react_1.useContext)(TimelineSelectionContext);
};
exports.useTimelineSelection = useTimelineSelection;
const useTimelineRowSelection = (nodePathInfo) => {
    const { canSelect, isSelected, selectItem } = (0, exports.useTimelineSelection)();
    const selectionItem = (0, react_1.useMemo)(() => nodePathInfo === null ? null : { type: 'row', nodePathInfo }, [nodePathInfo]);
    const selected = selectionItem === null ? false : isSelected(selectionItem);
    const onSelect = (0, react_1.useCallback)(() => {
        if (selectionItem === null) {
            return;
        }
        selectItem(selectionItem);
    }, [selectItem, selectionItem]);
    return {
        onSelect,
        selectable: canSelect && selectionItem !== null,
        selected,
    };
};
exports.useTimelineRowSelection = useTimelineRowSelection;
const useTimelineKeyframeSelection = (nodePathInfo, frame) => {
    const { canSelect, isSelected, selectItem } = (0, exports.useTimelineSelection)();
    const selectionItem = (0, react_1.useMemo)(() => ({
        type: 'keyframe',
        nodePathInfo,
        frame,
    }), [nodePathInfo, frame]);
    const selected = isSelected(selectionItem);
    const onSelect = (0, react_1.useCallback)(() => {
        selectItem(selectionItem);
    }, [selectItem, selectionItem]);
    return {
        onSelect,
        selectable: canSelect,
        selected,
    };
};
exports.useTimelineKeyframeSelection = useTimelineKeyframeSelection;
const useTimelineRowContainsSelection = (nodePathInfo) => {
    const { containsSelection } = (0, exports.useTimelineSelection)();
    if (nodePathInfo === null) {
        return false;
    }
    return containsSelection(nodePathInfo);
};
exports.useTimelineRowContainsSelection = useTimelineRowContainsSelection;
