// @remotion/studio/dist/components/Timeline/TimelineEffectGroupRow.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TimelineEffectGroupRow = void 0;
const jsx_runtime_1 = require("react/jsx-runtime");
const react_1 = require("react");
const remotion_1 = require("remotion");
const client_id_1 = require("../../helpers/client-id");
const timeline_layout_1 = require("../../helpers/timeline-layout");
const call_api_1 = require("../call-api");
const ContextMenu_1 = require("../ContextMenu");
const NotificationCenter_1 = require("../Notifications/NotificationCenter");
const save_effect_prop_1 = require("./save-effect-prop");
const TimelineExpandArrowButton_1 = require("./TimelineExpandArrowButton");
const TimelineLayerEye_1 = require("./TimelineLayerEye");
const TimelineRowChrome_1 = require("./TimelineRowChrome");
const TimelineSelection_1 = require("./TimelineSelection");
const rowLabel = {
    fontSize: 12,
    color: 'rgba(255, 255, 255, 0.8)',
    userSelect: 'none',
};
const TimelineEffectGroupRow = ({ label, nodePathInfo, effectIndex, effectSchema, documentationLink, nodePath, validatedLocation, rowDepth, getIsExpanded, toggleTrack, }) => {
    var _a;
    var _b;
    const [labelHovered, setLabelHovered] = (0, react_1.useState)(false);
    const { previewServerState } = (0, react_1.useContext)(client_id_1.StudioServerConnectionCtx);
    const previewConnected = previewServerState.type === 'connected';
    const { codeValues } = (0, react_1.useContext)(remotion_1.Internals.VisualModeCodeValuesContext);
    const { setCodeValues } = (0, react_1.useContext)(remotion_1.Internals.VisualModeSettersContext);
    const selection = (0, TimelineSelection_1.useTimelineRowSelection)(nodePathInfo);
    const effectStatus = (0, react_1.useMemo)(() => remotion_1.Internals.getEffectCodeValuesCtx({
        codeValues,
        nodePath,
        effectIndex,
    }), [codeValues, nodePath, effectIndex]);
    const disabledStatus = effectStatus.type === 'can-update-effect'
        ? ((_b = (_a = effectStatus.props) === null || _a === void 0 ? void 0 : _a.disabled) !== null && _b !== void 0 ? _b : null)
        : null;
    const isDisabled = (0, react_1.useMemo)(() => {
        if (disabledStatus && disabledStatus.canUpdate) {
            return Boolean(disabledStatus.codeValue);
        }
        return false;
    }, [disabledStatus]);
    const canToggle = previewConnected && disabledStatus !== null && disabledStatus.canUpdate;
    const deleteDisabled = !previewConnected ||
        effectStatus.type !== 'can-update-effect' ||
        !validatedLocation.source;
    const onDeleteEffectFromSource = (0, react_1.useCallback)(async () => {
        if (deleteDisabled) {
            return;
        }
        try {
            const result = await (0, call_api_1.callApi)('/api/delete-effect', {
                fileName: validatedLocation.source,
                sequenceNodePath: nodePath,
                effectIndex,
            });
            if (result.success) {
                (0, NotificationCenter_1.showNotification)('Removed effect from source file', 2000);
            }
            else {
                (0, NotificationCenter_1.showNotification)(result.reason, 4000);
            }
        }
        catch (err) {
            (0, NotificationCenter_1.showNotification)(err.message, 4000);
        }
    }, [deleteDisabled, effectIndex, nodePath, validatedLocation.source]);
    const contextMenuValues = (0, react_1.useMemo)(() => {
        if (!previewConnected) {
            return [];
        }
        return [
            {
                type: 'item',
                id: 'delete-effect',
                keyHint: null,
                label: 'Delete',
                leftItem: null,
                disabled: deleteDisabled,
                onClick: () => {
                    if (deleteDisabled) {
                        return;
                    }
                    onDeleteEffectFromSource();
                },
                quickSwitcherLabel: null,
                subMenu: null,
                value: 'delete-effect',
            },
        ];
    }, [deleteDisabled, onDeleteEffectFromSource, previewConnected]);
    const onToggle = (0, react_1.useCallback)((type) => {
        if (!canToggle || previewServerState.type !== 'connected') {
            return;
        }
        const newValue = type !== 'enable';
        const fieldSchema = effectSchema.disabled;
        const defaultValue = fieldSchema && fieldSchema.type === 'boolean'
            ? JSON.stringify(fieldSchema.default)
            : null;
        (0, save_effect_prop_1.saveEffectProp)({
            fileName: validatedLocation.source,
            nodePath,
            effectIndex,
            fieldKey: 'disabled',
            value: newValue,
            defaultValue,
            schema: effectSchema,
            setCodeValues,
            clientId: previewServerState.clientId,
        });
    }, [
        canToggle,
        effectIndex,
        effectSchema,
        nodePath,
        previewServerState,
        setCodeValues,
        validatedLocation.source,
    ]);
    const isExpanded = getIsExpanded(nodePathInfo);
    const rowStyle = (0, react_1.useMemo)(() => ({
        height: timeline_layout_1.TREE_GROUP_ROW_HEIGHT,
    }), []);
    const labelStyle = (0, react_1.useMemo)(() => {
        const hoverEffect = labelHovered && documentationLink !== null;
        return {
            ...rowLabel,
            ...(0, TimelineSelection_1.getTimelineSelectedLabelStyle)(selection.selected, true),
            alignSelf: 'stretch',
            alignItems: 'center',
            color: (0, TimelineSelection_1.getTimelineColor)(selection.selected, true),
            display: 'flex',
            flex: 1,
            minWidth: 0,
            paddingRight: timeline_layout_1.EXPANDED_SECTION_PADDING_RIGHT,
            textDecoration: hoverEffect ? 'underline' : 'none',
            textUnderlineOffset: 2,
            cursor: hoverEffect ? 'pointer' : undefined,
        };
    }, [documentationLink, labelHovered, selection.selected]);
    const onClickLabel = (0, react_1.useCallback)(() => {
        if (documentationLink === null) {
            return;
        }
        window.open(documentationLink, '_blank', 'noopener,noreferrer');
    }, [documentationLink]);
    const row = (jsx_runtime_1.jsx(TimelineRowChrome_1.TimelineRowChrome, { depth: rowDepth, eye: canToggle ? (jsx_runtime_1.jsx(TimelineLayerEye_1.TimelineLayerEye, { type: "effect", hidden: isDisabled, onInvoked: onToggle })) : (jsx_runtime_1.jsx(TimelineLayerEye_1.TimelineLayerEyeSpacer, {})), arrow: jsx_runtime_1.jsx(TimelineExpandArrowButton_1.TimelineExpandArrowButton, { isExpanded: isExpanded, onClick: () => toggleTrack(nodePathInfo), label: `${label} section`, disabled: false }), style: rowStyle, selected: selection.selected, selectable: selection.selectable, onSelect: selection.onSelect, showSelectedBackground: true, containsSelection: false, outerHeight: null, children: jsx_runtime_1.jsx("span", { onPointerEnter: () => setLabelHovered(true), onPointerLeave: () => setLabelHovered(false), onClick: onClickLabel, title: documentationLink ? `Open documentation: ${documentationLink}` : label, style: labelStyle, children: label }) }));
    return previewConnected ? (jsx_runtime_1.jsx(ContextMenu_1.ContextMenu, { values: contextMenuValues, onOpen: selection.selectable ? selection.onSelect : null, children: row })) : (row);
};
exports.TimelineEffectGroupRow = TimelineEffectGroupRow;
