// @remotion/studio/dist/components/Timeline/TimelineItemStack.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TimelineItemStack = void 0;
const jsx_runtime_1 = require("react/jsx-runtime");
const react_1 = require("react");
const client_id_1 = require("../../helpers/client-id");
const colors_1 = require("../../helpers/colors");
const get_git_menu_item_1 = require("../../helpers/get-git-menu-item");
const open_in_editor_1 = require("../../helpers/open-in-editor");
const layout_1 = require("../layout");
const NotificationCenter_1 = require("../Notifications/NotificationCenter");
const Spinner_1 = require("../Spinner");
const TimelineSelection_1 = require("./TimelineSelection");
const source_attribution_1 = require("./TimelineStack/source-attribution");
const TimelineItemStack = ({ originalLocation }) => {
    const [hovered, setHovered] = (0, react_1.useState)(false);
    const [opening, setOpening] = (0, react_1.useState)(false);
    const connectionStatus = (0, react_1.useContext)(client_id_1.StudioServerConnectionCtx)
        .previewServerState.type;
    const openEditor = (0, react_1.useCallback)(async (location) => {
        if (!window.remotion_editorName) {
            return;
        }
        setOpening(true);
        try {
            await (0, open_in_editor_1.openOriginalPositionInEditor)(location);
        }
        catch (err) {
            (0, NotificationCenter_1.showNotification)(err.message, 2000);
        }
        finally {
            setOpening(false);
        }
    }, []);
    const canOpenInEditor = Boolean(window.remotion_editorName &&
        connectionStatus === 'connected' &&
        originalLocation);
    const canOpenInGitHub = Boolean(window.remotion_gitSource && originalLocation);
    const hoverable = !TimelineSelection_1.SELECTION_ENABLED && (canOpenInEditor || canOpenInGitHub);
    const onClick = (0, react_1.useCallback)(() => {
        if (!originalLocation) {
            return;
        }
        if (canOpenInEditor) {
            openEditor(originalLocation);
            return;
        }
        if (canOpenInGitHub) {
            window.open((0, get_git_menu_item_1.getGitRefUrl)(window.remotion_gitSource, originalLocation), '_blank');
        }
    }, [canOpenInEditor, canOpenInGitHub, openEditor, originalLocation]);
    const onPointerEnter = (0, react_1.useCallback)(() => setHovered(true), []);
    const onPointerLeave = (0, react_1.useCallback)(() => setHovered(false), []);
    const style = (0, react_1.useMemo)(() => {
        return {
            fontSize: 12,
            color: opening
                ? colors_1.VERY_LIGHT_TEXT
                : hovered && hoverable
                    ? colors_1.LIGHT_TEXT
                    : colors_1.VERY_LIGHT_TEXT,
            cursor: hoverable ? 'pointer' : undefined,
            display: 'flex',
            flexDirection: 'row',
            alignItems: 'center',
            whiteSpace: 'nowrap',
            textOverflow: 'ellipsis',
            overflow: 'hidden',
            flexShrink: 100000,
            userSelect: 'none',
            WebkitUserSelect: 'none',
        };
    }, [opening, hovered, hoverable]);
    if (!originalLocation) {
        return null;
    }
    return (jsx_runtime_1.jsxs(jsx_runtime_1.Fragment, { children: [
            jsx_runtime_1.jsx("div", { onPointerEnter: hoverable ? onPointerEnter : undefined, onPointerLeave: hoverable ? onPointerLeave : undefined, onClick: hoverable ? onClick : undefined, style: style, children: (0, source_attribution_1.getOriginalSourceAttribution)(originalLocation) }), opening ? (jsx_runtime_1.jsxs(jsx_runtime_1.Fragment, { children: [
                    jsx_runtime_1.jsx(layout_1.Spacing, { x: 0.5 }), jsx_runtime_1.jsx(Spinner_1.Spinner, { duration: 0.5, size: 12 })
                ] })) : null] }));
};
exports.TimelineItemStack = TimelineItemStack;
