// @remotion/studio/dist/components/Timeline/TimelineSchemaField.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TimelineFieldValue = exports.TimelineNonEditableStatus = exports.UnsupportedStatus = void 0;
const jsx_runtime_1 = require("react/jsx-runtime");
const get_timeline_keyframes_1 = require("./get-timeline-keyframes");
const TimelineBooleanField_1 = require("./TimelineBooleanField");
const TimelineColorField_1 = require("./TimelineColorField");
const TimelineEnumField_1 = require("./TimelineEnumField");
const TimelineNumberField_1 = require("./TimelineNumberField");
const TimelineRotationField_1 = require("./TimelineRotationField");
const TimelineTranslateField_1 = require("./TimelineTranslateField");
const TimelineUvCoordinateField_1 = require("./TimelineUvCoordinateField");
const unsupportedLabel = {
    color: 'rgba(255, 255, 255, 0.4)',
    fontSize: 12,
    fontStyle: 'italic',
    userSelect: 'none',
    WebkitUserSelect: 'none',
};
const notEditableBackground = {
    backgroundColor: 'rgba(255, 0, 0, 0.2)',
    borderRadius: 3,
    padding: '0 4px',
};
const inlineWrapper = {
    fontSize: 12,
};
const UnsupportedStatus = ({ label }) => {
    return jsx_runtime_1.jsx("span", { style: unsupportedLabel, children: label });
};
exports.UnsupportedStatus = UnsupportedStatus;
const TimelineNonEditableStatus = ({ propStatus }) => {
    if (propStatus.canUpdate) {
        return null;
    }
    if (propStatus.reason === 'computed' || propStatus.reason === 'keyframed') {
        return (jsx_runtime_1.jsx("span", { style: unsupportedLabel, children: (0, get_timeline_keyframes_1.getComputedStatusLabel)(propStatus) }));
    }
};
exports.TimelineNonEditableStatus = TimelineNonEditableStatus;
const TimelineFieldValue = ({ field, onSave, onDragValueChange, onDragEnd, propStatus, effectiveValue, }) => {
    const wrapperStyle = !propStatus.canUpdate
        ? notEditableBackground
        : undefined;
    if (field.typeName === 'number') {
        return (jsx_runtime_1.jsx("span", { style: wrapperStyle, children: jsx_runtime_1.jsx(TimelineNumberField_1.TimelineNumberField, { field: field, effectiveValue: effectiveValue, onSave: onSave, propStatus: propStatus, onDragValueChange: onDragValueChange, onDragEnd: onDragEnd }) }));
    }
    if (field.typeName === 'rotation') {
        return (jsx_runtime_1.jsx("span", { style: wrapperStyle, children: jsx_runtime_1.jsx(TimelineRotationField_1.TimelineRotationField, { field: field, effectiveValue: effectiveValue, propStatus: propStatus, onSave: onSave, onDragValueChange: onDragValueChange, onDragEnd: onDragEnd }) }));
    }
    if (field.typeName === 'translate') {
        return (jsx_runtime_1.jsx("span", { style: wrapperStyle, children: jsx_runtime_1.jsx(TimelineTranslateField_1.TimelineTranslateField, { field: field, effectiveValue: effectiveValue, propStatus: propStatus, onSave: onSave, onDragValueChange: onDragValueChange, onDragEnd: onDragEnd }) }));
    }
    if (field.typeName === 'uv-coordinate') {
        return (jsx_runtime_1.jsx("span", { style: wrapperStyle, children: jsx_runtime_1.jsx(TimelineUvCoordinateField_1.TimelineUvCoordinateField, { field: field, effectiveValue: effectiveValue, propStatus: propStatus, onSave: onSave, onDragValueChange: onDragValueChange, onDragEnd: onDragEnd }) }));
    }
    if (field.typeName === 'boolean') {
        return (jsx_runtime_1.jsx("span", { style: wrapperStyle, children: jsx_runtime_1.jsx(TimelineBooleanField_1.TimelineBooleanField, { field: field, propStatus: propStatus, onSave: onSave, effectiveValue: effectiveValue }) }));
    }
    if (field.typeName === 'color') {
        return (jsx_runtime_1.jsx("span", { style: wrapperStyle, children: jsx_runtime_1.jsx(TimelineColorField_1.TimelineColorField, { field: field, propStatus: propStatus, onSave: onSave, onDragValueChange: onDragValueChange, onDragEnd: onDragEnd, effectiveValue: effectiveValue }) }));
    }
    if (field.typeName === 'enum') {
        return (jsx_runtime_1.jsx("span", { style: inlineWrapper, children: jsx_runtime_1.jsx(TimelineEnumField_1.TimelineEnumField, { field: field, propStatus: propStatus, onSave: onSave, effectiveValue: effectiveValue, onDragValueChange: onDragValueChange, onDragEnd: onDragEnd }) }));
    }
    throw new Error(`Unsupported field type: ${field.typeName}`);
};
exports.TimelineFieldValue = TimelineFieldValue;
