// @remotion/studio/dist/components/Timeline/TimelineListItem.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TimelineListItem = void 0;
const jsx_runtime_1 = require("react/jsx-runtime");
const react_1 = require("react");
const remotion_1 = require("remotion");
const no_react_1 = require("remotion/no-react");
const client_id_1 = require("../../helpers/client-id");
const timeline_layout_1 = require("../../helpers/timeline-layout");
const call_api_1 = require("../call-api");
const ContextMenu_1 = require("../ContextMenu");
const ExpandedTracksProvider_1 = require("../ExpandedTracksProvider");
const NotificationCenter_1 = require("../Notifications/NotificationCenter");
const save_sequence_prop_1 = require("./save-sequence-prop");
const TimelineExpandArrowButton_1 = require("./TimelineExpandArrowButton");
const TimelineExpandedSection_1 = require("./TimelineExpandedSection");
const TimelineItemStack_1 = require("./TimelineItemStack");
const TimelineLayerEye_1 = require("./TimelineLayerEye");
const TimelineMediaInfo_1 = require("./TimelineMediaInfo");
const TimelineRowChrome_1 = require("./TimelineRowChrome");
const TimelineSelection_1 = require("./TimelineSelection");
const TimelineSequenceName_1 = require("./TimelineSequenceName");
const use_open_sequence_in_editor_1 = require("./use-open-sequence-in-editor");
const labelContainerStyle = {
    alignItems: 'center',
    alignSelf: 'stretch',
    display: 'flex',
    flexDirection: 'row',
    minWidth: 0,
    gap: 4,
};
const TimelineListItem = ({ nestedDepth, sequence, nodePathInfo }) => {
    var _a;
    var _b;
    const nodePath = (_b = nodePathInfo === null || nodePathInfo === void 0 ? void 0 : nodePathInfo.sequenceSubscriptionKey) !== null && _b !== void 0 ? _b : null;
    const { previewServerState } = (0, react_1.useContext)(client_id_1.StudioServerConnectionCtx);
    const previewConnected = previewServerState.type === 'connected';
    const { getIsExpanded } = (0, react_1.useContext)(ExpandedTracksProvider_1.ExpandedTracksGetterContext);
    const { toggleTrack } = (0, react_1.useContext)(ExpandedTracksProvider_1.ExpandedTracksSetterContext);
    const { codeValues } = (0, react_1.useContext)(remotion_1.Internals.VisualModeCodeValuesContext);
    const { setCodeValues } = (0, react_1.useContext)(remotion_1.Internals.VisualModeSettersContext);
    const { setCanvasContent } = (0, react_1.useContext)(remotion_1.Internals.CompositionSetters);
    const { onSelect, selectable, selected } = (0, TimelineSelection_1.useTimelineRowSelection)(nodePathInfo);
    const containsSelection = (0, TimelineSelection_1.useTimelineRowContainsSelection)(nodePathInfo);
    const { canOpenInEditor, openInEditor, originalLocation } = (0, use_open_sequence_in_editor_1.useOpenSequenceInEditor)(sequence);
    const validatedLocation = (0, react_1.useMemo)(() => {
        var _a;
        if (!originalLocation ||
            !originalLocation.source ||
            !originalLocation.line) {
            return null;
        }
        return {
            source: originalLocation.source,
            line: originalLocation.line,
            column: (_a = originalLocation.column) !== null && _a !== void 0 ? _a : 0,
        };
    }, [originalLocation]);
    const canDeleteFromSource = Boolean(nodePath && (validatedLocation === null || validatedLocation === void 0 ? void 0 : validatedLocation.source));
    const deleteDisabled = (0, react_1.useMemo)(() => !previewConnected || !sequence.controls || !canDeleteFromSource, [previewConnected, sequence.controls, canDeleteFromSource]);
    const duplicateDisabled = deleteDisabled;
    const onDuplicateSequenceFromSource = (0, react_1.useCallback)(async () => {
        if (!(validatedLocation === null || validatedLocation === void 0 ? void 0 : validatedLocation.source) || !nodePath) {
            return;
        }
        if (nodePathInfo && nodePathInfo.numberOfSequencesWithThisNodePath > 1) {
            const message = 'This sequence is programmatically duplicated ' +
                nodePathInfo.numberOfSequencesWithThisNodePath +
                ' times in the code. Duplicating inserts another copy. Continue?';
            // eslint-disable-next-line no-alert -- native confirm before applying duplicate codemod in .map callbacks
            if (!window.confirm(message)) {
                return;
            }
        }
        try {
            const result = await (0, call_api_1.callApi)('/api/duplicate-jsx-node', {
                fileName: validatedLocation.source,
                nodePath: nodePath.nodePath,
            });
            if (result.success) {
                (0, NotificationCenter_1.showNotification)('Duplicated sequence in source file', 2000);
            }
            else {
                (0, NotificationCenter_1.showNotification)(result.reason, 4000);
            }
        }
        catch (err) {
            (0, NotificationCenter_1.showNotification)(err.message, 4000);
        }
    }, [nodePath, validatedLocation === null || validatedLocation === void 0 ? void 0 : validatedLocation.source, nodePathInfo]);
    const onDeleteSequenceFromSource = (0, react_1.useCallback)(async () => {
        if (!(validatedLocation === null || validatedLocation === void 0 ? void 0 : validatedLocation.source) || !nodePath) {
            return;
        }
        if (nodePathInfo && nodePathInfo.numberOfSequencesWithThisNodePath > 1) {
            const message = 'This sequence is programmatically duplicated ' +
                nodePathInfo.numberOfSequencesWithThisNodePath +
                ' times in the code. Deleting removes all instances. Continue?';
            // eslint-disable-next-line no-alert -- native confirm before applying duplicate codemod in .map callbacks
            if (!window.confirm(message)) {
                return;
            }
        }
        try {
            const result = await (0, call_api_1.callApi)('/api/delete-jsx-node', {
                fileName: validatedLocation.source,
                nodePath: nodePath.nodePath,
            });
            if (result.success) {
                (0, NotificationCenter_1.showNotification)('Removed sequence from source file', 2000);
            }
            else {
                (0, NotificationCenter_1.showNotification)(result.reason, 4000);
            }
        }
        catch (err) {
            (0, NotificationCenter_1.showNotification)(err.message, 4000);
        }
    }, [nodePath, validatedLocation === null || validatedLocation === void 0 ? void 0 : validatedLocation.source, nodePathInfo]);
    const mediaSrc = sequence.type === 'audio' ||
        sequence.type === 'video' ||
        sequence.type === 'image'
        ? sequence.src
        : null;
    const assetLinkInfo = (0, react_1.useMemo)(() => (mediaSrc ? (0, TimelineMediaInfo_1.getTimelineAssetLinkInfo)(mediaSrc) : null), [mediaSrc]);
    const contextMenuValues = (0, react_1.useMemo)(() => {
        if (!previewConnected) {
            return [];
        }
        const editorName = window.remotion_editorName;
        const { documentationLink } = sequence;
        return [
            editorName
                ? {
                    type: 'item',
                    id: 'show-in-editor',
                    keyHint: null,
                    label: `Show in ${editorName}`,
                    leftItem: null,
                    disabled: !canOpenInEditor,
                    onClick: () => {
                        openInEditor();
                    },
                    quickSwitcherLabel: null,
                    subMenu: null,
                    value: 'show-in-editor',
                }
                : null,
            documentationLink
                ? {
                    type: 'item',
                    id: 'open-component-docs',
                    keyHint: null,
                    label: 'Open component docs',
                    leftItem: null,
                    disabled: false,
                    onClick: () => {
                        window.open(documentationLink, '_blank', 'noopener,noreferrer');
                    },
                    quickSwitcherLabel: null,
                    subMenu: null,
                    value: 'open-component-docs',
                }
                : null,
            assetLinkInfo
                ? {
                    type: 'item',
                    id: 'show-asset',
                    keyHint: null,
                    label: 'Show asset',
                    leftItem: null,
                    disabled: false,
                    onClick: () => {
                        (0, TimelineMediaInfo_1.openTimelineAssetLink)(assetLinkInfo, setCanvasContent);
                    },
                    quickSwitcherLabel: null,
                    subMenu: null,
                    value: 'show-asset',
                }
                : null,
            documentationLink
                ? {
                    type: 'divider',
                    id: 'open-component-docs-divider',
                }
                : null,
            {
                type: 'item',
                id: 'duplicate-sequence',
                keyHint: null,
                label: 'Duplicate',
                leftItem: null,
                disabled: duplicateDisabled,
                onClick: () => {
                    if (duplicateDisabled) {
                        return;
                    }
                    onDuplicateSequenceFromSource();
                },
                quickSwitcherLabel: null,
                subMenu: null,
                value: 'duplicate-sequence',
            },
            {
                type: 'item',
                id: 'delete-sequence',
                keyHint: null,
                label: 'Delete',
                leftItem: null,
                disabled: deleteDisabled,
                onClick: () => {
                    if (deleteDisabled) {
                        return;
                    }
                    onDeleteSequenceFromSource();
                },
                quickSwitcherLabel: null,
                subMenu: null,
                value: 'delete-sequence',
            },
        ].filter(no_react_1.NoReactInternals.truthy);
    }, [
        assetLinkInfo,
        deleteDisabled,
        duplicateDisabled,
        onDeleteSequenceFromSource,
        onDuplicateSequenceFromSource,
        canOpenInEditor,
        openInEditor,
        previewConnected,
        sequence,
        setCanvasContent,
    ]);
    const isExpanded = previewConnected && nodePathInfo !== null && getIsExpanded(nodePathInfo);
    const onToggleExpand = (0, react_1.useCallback)(() => {
        if (nodePathInfo === null) {
            return;
        }
        toggleTrack(nodePathInfo);
    }, [nodePathInfo, toggleTrack]);
    const onShowInEditorDoubleClick = (0, react_1.useCallback)((e) => {
        if (!TimelineSelection_1.SELECTION_ENABLED || !canOpenInEditor) {
            return;
        }
        e.stopPropagation();
        openInEditor();
    }, [canOpenInEditor, openInEditor]);
    const codeValuesForOverride = (0, react_1.useMemo)(() => {
        return nodePath
            ? remotion_1.Internals.getCodeValuesCtx(codeValues, nodePath)
            : undefined;
    }, [codeValues, nodePath]);
    const codeHiddenStatus = codeValuesForOverride === null || codeValuesForOverride === void 0 ? void 0 : codeValuesForOverride.hidden;
    const isItemHidden = (0, react_1.useMemo)(() => {
        var _a;
        const codeValue = codeHiddenStatus && codeHiddenStatus.canUpdate
            ? codeHiddenStatus.codeValue
            : undefined;
        const runtimeValue = (_a = sequence.controls) === null || _a === void 0 ? void 0 : _a.currentRuntimeValueDotNotation.hidden;
        const effective = (codeValue !== null && codeValue !== void 0 ? codeValue : runtimeValue);
        return effective !== null && effective !== void 0 ? effective : false;
    }, [codeHiddenStatus, (_a = sequence.controls) === null || _a === void 0 ? void 0 : _a.currentRuntimeValueDotNotation]);
    const onToggleVisibility = (0, react_1.useCallback)((type) => {
        if (!sequence.controls ||
            !nodePath ||
            !validatedLocation ||
            !codeValuesForOverride ||
            !codeHiddenStatus ||
            !codeHiddenStatus.canUpdate ||
            previewServerState.type !== 'connected') {
            return;
        }
        const newValue = type !== 'enable';
        const { schema } = sequence.controls;
        const fieldSchema = schema.hidden;
        const defaultValue = fieldSchema && fieldSchema.type === 'boolean'
            ? JSON.stringify(fieldSchema.default)
            : null;
        (0, save_sequence_prop_1.saveSequenceProp)({
            fileName: validatedLocation.source,
            nodePath,
            fieldKey: 'hidden',
            value: newValue,
            defaultValue,
            schema,
            setCodeValues,
            clientId: previewServerState.clientId,
        });
    }, [
        codeHiddenStatus,
        codeValuesForOverride,
        nodePath,
        previewServerState,
        sequence.controls,
        setCodeValues,
        validatedLocation,
    ]);
    const outerHeight = (0, react_1.useMemo)(() => (0, timeline_layout_1.getTimelineLayerHeight)(sequence.type) + timeline_layout_1.TIMELINE_ITEM_BORDER_BOTTOM, [sequence.type]);
    const inner = (0, react_1.useMemo)(() => {
        return {
            height: timeline_layout_1.TIMELINE_LIST_ITEM_ROW_HEIGHT,
            color: 'white',
            fontFamily: 'Arial, Helvetica, sans-serif',
            wordBreak: 'break-all',
            textAlign: 'left',
            flexShrink: 0,
        };
    }, []);
    const hasExpandableContent = Boolean(sequence.controls) || sequence.effects.length > 0;
    const canToggleVisibility = previewConnected &&
        Boolean(sequence.controls) &&
        nodePath !== null &&
        validatedLocation !== null &&
        codeHiddenStatus !== undefined &&
        codeHiddenStatus !== null &&
        codeHiddenStatus.canUpdate;
    const trackRow = (jsx_runtime_1.jsx(TimelineRowChrome_1.TimelineRowChrome, { depth: nestedDepth, eye: canToggleVisibility ? (jsx_runtime_1.jsx(TimelineLayerEye_1.TimelineLayerEye, { type: sequence.type === 'audio' ? 'speaker' : 'eye', hidden: isItemHidden, onInvoked: onToggleVisibility })) : (jsx_runtime_1.jsx(TimelineLayerEye_1.TimelineLayerEyeSpacer, {})), arrow: hasExpandableContent ? (jsx_runtime_1.jsx(TimelineExpandArrowButton_1.TimelineExpandArrowButton, { isExpanded: isExpanded, onClick: onToggleExpand, label: "track properties", disabled: !previewConnected || nodePathInfo === null })) : (jsx_runtime_1.jsx(TimelineExpandArrowButton_1.TimelineExpandArrowSpacer, {})), style: inner, selected: selected, selectable: selectable, onSelect: onSelect, showSelectedBackground: true, containsSelection: containsSelection, outerHeight: outerHeight, onDoubleClick: TimelineSelection_1.SELECTION_ENABLED && canOpenInEditor
            ? onShowInEditorDoubleClick
            : undefined, children: jsx_runtime_1.jsxs("div", { style: labelContainerStyle, children: [
                jsx_runtime_1.jsx(TimelineSequenceName_1.TimelineSequenceName, { sequence: sequence, selected: selected, containsSelection: containsSelection }), mediaSrc ? jsx_runtime_1.jsx(TimelineMediaInfo_1.TimelineMediaInfo, { src: mediaSrc }) : null, jsx_runtime_1.jsx(TimelineItemStack_1.TimelineItemStack, { originalLocation: originalLocation })
            ] }) }));
    return (jsx_runtime_1.jsxs(jsx_runtime_1.Fragment, { children: [previewConnected ? (jsx_runtime_1.jsx(ContextMenu_1.ContextMenu, { values: contextMenuValues, onOpen: selectable ? onSelect : null, children: trackRow })) : (trackRow), previewConnected &&
                isExpanded &&
                hasExpandableContent &&
                nodePathInfo &&
                validatedLocation ? (jsx_runtime_1.jsx(TimelineExpandedSection_1.TimelineExpandedSection, { sequence: sequence, validatedLocation: validatedLocation, nodePathInfo: nodePathInfo, nestedDepth: nestedDepth })) : null] }));
};
exports.TimelineListItem = TimelineListItem;
