// @remotion/studio/dist/components/Timeline/TimelineEffectFieldRow.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TimelineEffectFieldRow = void 0;
const jsx_runtime_1 = require("react/jsx-runtime");
const studio_shared_1 = require("@remotion/studio-shared");
const react_1 = require("react");
const remotion_1 = require("remotion");
const client_id_1 = require("../../helpers/client-id");
const call_api_1 = require("../call-api");
const get_timeline_keyframes_1 = require("./get-timeline-keyframes");
const save_prop_queue_1 = require("./save-prop-queue");
const timeline_field_row_layout_1 = require("./timeline-field-row-layout");
const TimelineExpandArrowButton_1 = require("./TimelineExpandArrowButton");
const TimelineFieldLabel_1 = require("./TimelineFieldLabel");
const TimelineLayerEye_1 = require("./TimelineLayerEye");
const TimelineRowChrome_1 = require("./TimelineRowChrome");
const TimelineSchemaField_1 = require("./TimelineSchemaField");
const TimelineSelection_1 = require("./TimelineSelection");
const fieldRowBase = {};
const Value = ({ field, nodePath, validatedLocation }) => {
    var _a;
    var _b;
    const { setEffectDragOverrides, clearEffectDragOverrides, setCodeValues } = (0, react_1.useContext)(remotion_1.Internals.VisualModeSettersContext);
    const { getEffectDragOverrides } = (0, react_1.useContext)(remotion_1.Internals.VisualModeDragOverridesContext);
    const { codeValues: visualModeCodeValues } = (0, react_1.useContext)(remotion_1.Internals.VisualModeCodeValuesContext);
    const { previewServerState } = (0, react_1.useContext)(client_id_1.StudioServerConnectionCtx);
    const clientId = previewServerState.type === 'connected'
        ? previewServerState.clientId
        : null;
    const effectStatus = remotion_1.Internals.getEffectCodeValuesCtx({
        codeValues: visualModeCodeValues,
        nodePath,
        effectIndex: field.effectIndex,
    });
    const propStatus = effectStatus.type === 'can-update-effect'
        ? ((_b = (_a = effectStatus.props) === null || _a === void 0 ? void 0 : _a[field.key]) !== null && _b !== void 0 ? _b : null)
        : null;
    const onDragValueChange = (0, react_1.useCallback)((value) => {
        setEffectDragOverrides(nodePath, field.effectIndex, field.key, value);
    }, [setEffectDragOverrides, nodePath, field.effectIndex, field.key]);
    const onDragEnd = (0, react_1.useCallback)(() => {
        clearEffectDragOverrides(nodePath, field.effectIndex);
    }, [clearEffectDragOverrides, nodePath, field.effectIndex]);
    const dragOverrideValue = (0, react_1.useMemo)(() => {
        const overrides = getEffectDragOverrides(nodePath, field.effectIndex);
        return overrides[field.key];
    }, [getEffectDragOverrides, nodePath, field.effectIndex, field.key]);
    const onSave = (0, react_1.useCallback)((value) => {
        if (!validatedLocation) {
            return Promise.reject(new Error('Cannot save'));
        }
        if (!propStatus) {
            return Promise.reject(new Error('Cannot save'));
        }
        if (!propStatus.canUpdate) {
            return Promise.reject(new Error('Cannot save'));
        }
        if (!clientId) {
            return Promise.reject(new Error('Not connected to studio server'));
        }
        const defaultValue = field.fieldSchema.default !== undefined
            ? JSON.stringify(field.fieldSchema.default)
            : null;
        const stringifiedValue = JSON.stringify(value);
        if (value === propStatus.codeValue) {
            return Promise.resolve();
        }
        if (defaultValue === stringifiedValue &&
            propStatus.codeValue === undefined) {
            return Promise.resolve();
        }
        return (0, save_prop_queue_1.enqueueSavePropChange)({
            nodePath,
            setCodeValues,
            applyOptimistic: (prev) => (0, studio_shared_1.optimisticUpdateForEffectCodeValues)({
                previous: prev,
                effectIndex: field.effectIndex,
                fieldKey: field.key,
                value,
                schema: field.effectSchema,
            }),
            apiCall: () => (0, call_api_1.callApi)('/api/save-effect-props', {
                fileName: validatedLocation.source,
                sequenceNodePath: nodePath,
                effectIndex: field.effectIndex,
                key: field.key,
                value: stringifiedValue,
                defaultValue,
                schema: field.effectSchema,
                clientId,
            }),
            errorLabel: 'Could not save effect prop',
        });
    }, [
        clientId,
        field.effectIndex,
        field.effectSchema,
        field.fieldSchema.default,
        field.key,
        nodePath,
        propStatus,
        setCodeValues,
        validatedLocation,
    ]);
    if (effectStatus.type === 'cannot-update-effect') {
        if (effectStatus.reason === 'computed') {
            return jsx_runtime_1.jsx(TimelineSchemaField_1.UnsupportedStatus, { label: "computed" });
        }
        if (effectStatus.reason === 'not-call-expression') {
            return jsx_runtime_1.jsx(TimelineSchemaField_1.UnsupportedStatus, { label: "not inline" });
        }
        if (effectStatus.reason === 'not-found') {
            return jsx_runtime_1.jsx(TimelineSchemaField_1.UnsupportedStatus, { label: "not found in code" });
        }
        throw new Error(`Unsupported effect status: ${effectStatus.reason}`);
    }
    if (effectStatus.type === 'cannot-update-sequence') {
        if (effectStatus.reason === 'not-found') {
            return jsx_runtime_1.jsx(TimelineSchemaField_1.UnsupportedStatus, { label: "not found in code" });
        }
        if (effectStatus.reason === 'error') {
            return jsx_runtime_1.jsx(TimelineSchemaField_1.UnsupportedStatus, { label: "error" });
        }
        throw new Error(`Unsupported effect status: ${effectStatus.reason}`);
    }
    if (propStatus === null || !propStatus.canUpdate) {
        if ((propStatus === null || propStatus === void 0 ? void 0 : propStatus.reason) === 'computed' ||
            (propStatus === null || propStatus === void 0 ? void 0 : propStatus.reason) === 'keyframed') {
            return jsx_runtime_1.jsx(TimelineSchemaField_1.UnsupportedStatus, { label: (0, get_timeline_keyframes_1.getComputedStatusLabel)(propStatus) });
        }
        return null;
    }
    const effectiveValue = remotion_1.Internals.getEffectiveVisualModeValue({
        codeValue: propStatus,
        dragOverrideValue,
        defaultValue: field.fieldSchema.default,
        shouldResortToDefaultValueIfUndefined: true,
    });
    return (jsx_runtime_1.jsx(TimelineSchemaField_1.TimelineFieldValue, { field: field, propStatus: propStatus, onSave: onSave, onDragValueChange: onDragValueChange, onDragEnd: onDragEnd, effectiveValue: effectiveValue }));
};
const TimelineEffectFieldRow = ({ field, validatedLocation, rowDepth, nodePath, nodePathInfo }) => {
    var _a;
    const selection = (0, TimelineSelection_1.useTimelineRowSelection)(nodePathInfo);
    const style = (0, react_1.useMemo)(() => {
        return {
            ...fieldRowBase,
            height: field.rowHeight,
        };
    }, [field.rowHeight]);
    return (jsx_runtime_1.jsxs(TimelineRowChrome_1.TimelineRowChrome, { depth: rowDepth, eye: jsx_runtime_1.jsx(TimelineLayerEye_1.TimelineLayerEyeSpacer, {}), arrow: jsx_runtime_1.jsx(TimelineExpandArrowButton_1.TimelineExpandArrowSpacer, {}), style: style, selected: selection.selected, selectable: selection.selectable, onSelect: selection.onSelect, showSelectedBackground: true, containsSelection: false, outerHeight: null, children: [
            jsx_runtime_1.jsx(TimelineFieldLabel_1.TimelineFieldLabel, { rowDepth: rowDepth, selected: selection.selected, label: (_a = field.description) !== null && _a !== void 0 ? _a : field.key }), jsx_runtime_1.jsx("div", { style: timeline_field_row_layout_1.timelineFieldValueColumnStyle, children: jsx_runtime_1.jsx(Value, { field: field, nodePath: nodePath, validatedLocation: validatedLocation }) })
        ] }));
};
exports.TimelineEffectFieldRow = TimelineEffectFieldRow;
