// @remotion/studio/dist/components/Timeline/use-expanded-track-keyframe-rows.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.useExpandedTrackKeyframeRows = void 0;
const react_1 = require("react");
const remotion_1 = require("remotion");
const timeline_layout_1 = require("../../helpers/timeline-layout");
const timeline_node_path_key_1 = require("../../helpers/timeline-node-path-key");
const ExpandedTracksProvider_1 = require("../ExpandedTracksProvider");
const get_node_keyframes_1 = require("./get-node-keyframes");
const useExpandedTrackKeyframeRows = ({ sequence, nodePathInfo, keyframeDisplayOffset, }) => {
    const { getIsExpanded } = (0, react_1.useContext)(ExpandedTracksProvider_1.ExpandedTracksGetterContext);
    const { codeValues } = (0, react_1.useContext)(remotion_1.Internals.VisualModeCodeValuesContext);
    const tree = (0, react_1.useMemo)(() => (0, timeline_layout_1.buildTimelineTree)({
        sequence,
        nodePathInfo,
        getDragOverrides: () => ({}),
        getEffectDragOverrides: () => ({}),
        codeValues,
    }), [codeValues, nodePathInfo, sequence]);
    const flat = (0, react_1.useMemo)(() => (0, timeline_layout_1.flattenVisibleTreeNodes)({ nodes: tree, getIsExpanded }), [tree, getIsExpanded]);
    const expandedHeight = (0, react_1.useMemo)(() => (0, timeline_layout_1.getExpandedTrackHeight)({
        sequence,
        nodePathInfo,
        getIsExpanded,
        codeValues,
    }), [codeValues, getIsExpanded, nodePathInfo, sequence]);
    const rows = (0, react_1.useMemo)(() => flat.map(({ node }) => ({
        height: (0, timeline_layout_1.getTreeRowHeight)(node),
        keyframes: (0, get_node_keyframes_1.getNodeKeyframes)({
            node,
            nodePath: nodePathInfo.sequenceSubscriptionKey,
            codeValues,
            keyframeDisplayOffset,
        }),
        rowKey: (0, timeline_node_path_key_1.timelineNodePathInfoToKey)(node.nodePathInfo),
        nodePathInfo: node.nodePathInfo,
    })), [
        codeValues,
        flat,
        keyframeDisplayOffset,
        nodePathInfo.sequenceSubscriptionKey,
    ]);
    return { rows, expandedHeight };
};
exports.useExpandedTrackKeyframeRows = useExpandedTrackKeyframeRows;
