// @remotion/studio-server/dist/preview-server/sequence-props-watchers.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.unsubscribeClientSequencePropsWatchers = exports.unsubscribeFromSequencePropsWatchers = exports.subscribeToSequencePropsWatchers = void 0;
const node_path_1 = __importDefault(require("node:path"));
const renderer_1 = require("@remotion/renderer");
const studio_shared_1 = require("@remotion/studio-shared");
const file_watcher_1 = require("../file-watcher");
const jsx_element_not_found_at_location_error_1 = require("./jsx-element-not-found-at-location-error");
const live_events_1 = require("./live-events");
const node_path_cache_1 = require("./node-path-cache");
const can_update_sequence_props_1 = require("./routes/can-update-sequence-props");
const sequencePropsWatchers = {};
const getSequencePropsStatus = ({ fileName, line, column, keys, effects, remotionRoot, logLevel, }) => {
    // Try cached nodePath first (handles stale source maps after suppressed rebuilds)
    const cachedNodePath = (0, node_path_cache_1.getCachedNodePath)(fileName, line, column);
    if (cachedNodePath) {
        const cachedResult = (0, can_update_sequence_props_1.computeSequencePropsStatus)({
            fileName,
            nodePath: cachedNodePath,
            keys,
            effects,
            remotionRoot,
        });
        if (cachedResult.canUpdate) {
            return {
                status: cachedResult,
                nodePath: {
                    absolutePath: node_path_1.default.resolve(remotionRoot, fileName),
                    nodePath: cachedNodePath,
                    sequenceKeys: keys,
                    effectKeys: effects,
                },
                success: true,
            };
        }
    }
    const status = (0, can_update_sequence_props_1.computeSequencePropsStatusFromFilenameByLine)({
        fileName,
        line,
        keys,
        effects,
        remotionRoot,
        logLevel,
    });
    return status;
};
const subscribeToSequencePropsWatchers = ({ fileName, line, column, keys, effects, remotionRoot, clientId, logLevel, }) => {
    var _a;
    const initialResult = getSequencePropsStatus({
        fileName,
        line,
        column,
        keys,
        effects,
        remotionRoot,
        logLevel,
    });
    if (!initialResult.success) {
        return initialResult;
    }
    const absolutePath = node_path_1.default.resolve(remotionRoot, fileName);
    // Cache the resolved nodePath for future lookups with stale source maps
    (0, node_path_cache_1.setCachedNodePath)(fileName, line, column, initialResult.nodePath.nodePath);
    const { nodePath } = initialResult;
    const watcherKey = (0, studio_shared_1.stringifySequenceSubscriptionKey)(nodePath);
    // If a watcher already exists for this key, just bump the ref count
    if ((_a = sequencePropsWatchers[clientId]) === null || _a === void 0 ? void 0 : _a[watcherKey]) {
        sequencePropsWatchers[clientId][watcherKey].refCount++;
        return initialResult;
    }
    const { unwatch } = (0, file_watcher_1.installFileWatcher)({
        file: absolutePath,
        existenceOnly: false,
        onChange: (event) => {
            if (event.type === 'deleted') {
                return;
            }
            if (event.originatorClientId === clientId) {
                return;
            }
            try {
                const result = (0, can_update_sequence_props_1.computeSequencePropsStatusFromContent)({
                    fileContents: event.content,
                    nodePath: nodePath.nodePath,
                    keys,
                    effects,
                });
                const previousEffectChain = result.effects.map((effect) => effect.canUpdate && effect.callee);
                const newEffectChain = initialResult.success &&
                    initialResult.status.effects.map((effect) => effect.canUpdate && effect.callee);
                if (previousEffectChain.join(',') !== newEffectChain.join(',')) {
                    renderer_1.RenderInternals.Log.verbose({ indent: false, logLevel }, 'Effect chain changed, not sending "sequence-props-updated" event');
                    return;
                }
                (0, live_events_1.waitForLiveEventsListener)().then((listener) => {
                    listener.sendEventToClientId(clientId, {
                        type: 'sequence-props-updated',
                        fileName,
                        nodePath,
                        result,
                    });
                });
            }
            catch (error) {
                if (error instanceof jsx_element_not_found_at_location_error_1.JsxElementNotFoundAtLocationError) {
                    (0, live_events_1.waitForLiveEventsListener)().then((listener) => {
                        listener.sendEventToClientId(clientId, {
                            type: 'lost-node-path',
                            fileName,
                            line,
                            column,
                        });
                    });
                    return;
                }
                renderer_1.RenderInternals.Log.error({ indent: false, logLevel }, error);
            }
        },
    });
    if (!sequencePropsWatchers[clientId]) {
        sequencePropsWatchers[clientId] = {};
    }
    sequencePropsWatchers[clientId][watcherKey] = { unwatch, refCount: 1 };
    return initialResult;
};
exports.subscribeToSequencePropsWatchers = subscribeToSequencePropsWatchers;
const unsubscribeFromSequencePropsWatchers = ({ fileName, nodePath, remotionRoot, clientId, sequenceKeys, effectKeys, }) => {
    const absolutePath = node_path_1.default.resolve(remotionRoot, fileName);
    const watcherKey = (0, studio_shared_1.stringifySequenceSubscriptionKey)({
        absolutePath,
        nodePath,
        sequenceKeys,
        effectKeys,
    });
    if (!sequencePropsWatchers[clientId] ||
        !sequencePropsWatchers[clientId][watcherKey]) {
        return;
    }
    sequencePropsWatchers[clientId][watcherKey].refCount--;
    if (sequencePropsWatchers[clientId][watcherKey].refCount <= 0) {
        sequencePropsWatchers[clientId][watcherKey].unwatch();
        delete sequencePropsWatchers[clientId][watcherKey];
    }
};
exports.unsubscribeFromSequencePropsWatchers = unsubscribeFromSequencePropsWatchers;
const unsubscribeClientSequencePropsWatchers = (clientId) => {
    if (!sequencePropsWatchers[clientId]) {
        return;
    }
    Object.values(sequencePropsWatchers[clientId]).forEach((watcher) => {
        watcher.unwatch();
    });
    delete sequencePropsWatchers[clientId];
};
exports.unsubscribeClientSequencePropsWatchers = unsubscribeClientSequencePropsWatchers;
