// @remotion/studio/dist/helpers/open-in-editor.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.openCompositionComponentInEditor = exports.preloadCompositionComponentInfo = exports.openOriginalPositionInEditor = exports.openInEditor = void 0;
const openInEditor = (stack) => {
    const { originalFileName, originalLineNumber, originalColumnNumber, originalFunctionName, originalScriptCode, } = stack;
    return fetch(`/api/open-in-editor`, {
        method: 'post',
        headers: {
            'content-type': 'application/json',
        },
        body: JSON.stringify({
            stack: {
                originalFileName,
                originalLineNumber,
                originalColumnNumber,
                originalFunctionName,
                originalScriptCode,
            },
        }),
    });
};
exports.openInEditor = openInEditor;
const openOriginalPositionInEditor = async (originalPosition) => {
    await (0, exports.openInEditor)({
        originalColumnNumber: originalPosition.column,
        originalFileName: originalPosition.source,
        originalFunctionName: null,
        originalLineNumber: originalPosition.line,
        originalScriptCode: null,
    });
};
exports.openOriginalPositionInEditor = openOriginalPositionInEditor;
const componentResolutionCache = new Map();
const getComponentResolutionCacheKey = ({ compositionFile, compositionId, }) => {
    return `${compositionFile}::${compositionId}`;
};
const loadCompositionComponentInfo = async ({ compositionFile, compositionId, }) => {
    const cacheKey = getComponentResolutionCacheKey({
        compositionFile,
        compositionId,
    });
    const existing = componentResolutionCache.get(cacheKey);
    if (existing) {
        return existing;
    }
    const promise = (async () => {
        const response = await fetch(`/api/composition-component-info`, {
            method: 'post',
            headers: {
                'content-type': 'application/json',
            },
            body: JSON.stringify({
                compositionFile,
                compositionId,
            }),
        });
        const body = (await response.json());
        if (!body.success) {
            throw new Error(body.error);
        }
        return {
            location: body.location,
            canAddSequence: body.canAddSequence,
        };
    })();
    componentResolutionCache.set(cacheKey, promise);
    try {
        return await promise;
    }
    catch (err) {
        componentResolutionCache.delete(cacheKey);
        throw err;
    }
};
const preloadCompositionComponentInfo = ({ compositionFile, compositionId, }) => {
    loadCompositionComponentInfo({
        compositionFile,
        compositionId,
    }).catch(() => undefined);
};
exports.preloadCompositionComponentInfo = preloadCompositionComponentInfo;
const openCompositionComponentInEditor = async ({ compositionFile, compositionId, }) => {
    const info = await loadCompositionComponentInfo({
        compositionFile,
        compositionId,
    });
    await (0, exports.openOriginalPositionInEditor)(info.location);
};
exports.openCompositionComponentInEditor = openCompositionComponentInEditor;
