// @remotion/studio-server/dist/codemods/update-keyframes/update-keyframes.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.updateEffectKeyframes = exports.updateEffectKeyframesAst = exports.updateSequenceKeyframes = exports.updateSequenceKeyframesAst = void 0;
const recast = __importStar(require("recast"));
const can_update_sequence_props_1 = require("../../preview-server/routes/can-update-sequence-props");
const format_file_content_1 = require("../format-file-content");
const parse_ast_1 = require("../parse-ast");
const update_effect_props_1 = require("../update-effect-props/update-effect-props");
const update_nested_prop_1 = require("../update-nested-prop");
const b = recast.types.builders;
const getSupportedCallArgument = (arg) => {
    if (arg === undefined ||
        arg.type === 'ArgumentPlaceholder' ||
        arg.type === 'JSXNamespacedName') {
        return null;
    }
    return arg;
};
const getNumericValue = (node) => {
    if (node.type === 'NumericLiteral') {
        return node.value;
    }
    if (node.type === 'UnaryExpression' &&
        (node.operator === '-' || node.operator === '+') &&
        node.argument.type === 'NumericLiteral') {
        return node.operator === '-' ? -node.argument.value : node.argument.value;
    }
    if (node.type === 'TSAsExpression') {
        return getNumericValue(node.expression);
    }
    return null;
};
const getInterpolationExpression = (node) => {
    if (node.type === 'TSAsExpression') {
        return getInterpolationExpression(node.expression);
    }
    if (node.type !== 'CallExpression' ||
        node.callee.type !== 'Identifier' ||
        (node.callee.name !== 'interpolate' &&
            node.callee.name !== 'interpolateColors')) {
        return null;
    }
    const frameArg = getSupportedCallArgument(node.arguments[0]);
    const inputArg = node.arguments[1];
    const outputArg = node.arguments[2];
    if (!frameArg ||
        !inputArg ||
        !outputArg ||
        inputArg.type !== 'ArrayExpression' ||
        outputArg.type !== 'ArrayExpression' ||
        inputArg.elements.length !== outputArg.elements.length) {
        return null;
    }
    const keyframes = [];
    for (let i = 0; i < inputArg.elements.length; i++) {
        const inputElement = inputArg.elements[i];
        const outputElement = outputArg.elements[i];
        if (!inputElement ||
            !outputElement ||
            inputElement.type === 'SpreadElement' ||
            outputElement.type === 'SpreadElement') {
            return null;
        }
        const frame = getNumericValue(inputElement);
        if (frame === null || !(0, can_update_sequence_props_1.isStaticValue)(outputElement)) {
            return null;
        }
        keyframes.push({
            frame,
            output: outputElement,
            value: (0, can_update_sequence_props_1.extractStaticValue)(outputElement),
        });
    }
    if (keyframes.length === 0) {
        return null;
    }
    const extraArgs = [];
    for (const arg of node.arguments.slice(3)) {
        const supportedArg = getSupportedCallArgument(arg);
        if (!supportedArg) {
            return null;
        }
        extraArgs.push(supportedArg);
    }
    return {
        callee: node.callee,
        input: frameArg,
        extraArgs,
        keyframes,
    };
};
const getInterpolationCalleeForValues = ({ staticValue, newValue, }) => {
    return b.identifier(typeof staticValue === 'string' && typeof newValue === 'string'
        ? 'interpolateColors'
        : 'interpolate');
};
const createFrameExpression = (frame) => {
    return (0, update_nested_prop_1.parseValueExpression)(frame);
};
const createInterpolateExpression = ({ callee, input, extraArgs, keyframes, }) => {
    const sortedKeyframes = [...keyframes].sort((first, second) => first.frame - second.frame);
    return b.callExpression(callee, [
        input,
        b.arrayExpression(sortedKeyframes.map((keyframe) => createFrameExpression(keyframe.frame))),
        b.arrayExpression(sortedKeyframes.map((keyframe) => keyframe.output)),
        ...extraArgs,
    ]);
};
const addKeyframe = ({ expression, frame, value, }) => {
    const existing = getInterpolationExpression(expression);
    const newOutput = (0, update_nested_prop_1.parseValueExpression)(value);
    if (existing) {
        const existingIndex = existing.keyframes.findIndex((keyframe) => keyframe.frame === frame);
        const nextKeyframes = existingIndex === -1
            ? [...existing.keyframes, { frame, output: newOutput, value }]
            : existing.keyframes.map((keyframe, index) => index === existingIndex
                ? { frame, output: newOutput, value }
                : keyframe);
        return createInterpolateExpression({
            callee: existing.callee,
            input: existing.input,
            extraArgs: existing.extraArgs,
            keyframes: nextKeyframes,
        });
    }
    if (!(0, can_update_sequence_props_1.isStaticValue)(expression)) {
        throw new Error('Cannot add keyframe to computed expression');
    }
    const staticValue = (0, can_update_sequence_props_1.extractStaticValue)(expression);
    const staticOutput = (0, update_nested_prop_1.parseValueExpression)(staticValue);
    const keyframes = frame === 0
        ? [{ frame, output: newOutput, value }]
        : [
            { frame: 0, output: staticOutput, value: staticValue },
            { frame, output: newOutput, value },
        ];
    return createInterpolateExpression({
        callee: getInterpolationCalleeForValues({
            staticValue,
            newValue: value,
        }),
        input: b.identifier('frame'),
        extraArgs: [],
        keyframes,
    });
};
const removeKeyframe = ({ expression, frame, }) => {
    const existing = getInterpolationExpression(expression);
    if (!existing) {
        throw new Error('Cannot remove keyframe from non-interpolated expression');
    }
    const keyframeIndex = existing.keyframes.findIndex((keyframe) => keyframe.frame === frame);
    if (keyframeIndex === -1) {
        throw new Error(`Cannot remove keyframe at frame ${frame}: not found`);
    }
    const nextKeyframes = existing.keyframes.filter((_keyframe, index) => index !== keyframeIndex);
    return createInterpolateExpression({
        callee: existing.callee,
        input: existing.input,
        extraArgs: existing.extraArgs,
        keyframes: nextKeyframes,
    });
};
const applyKeyframeOperation = ({ expression, operation, }) => {
    if (operation.type === 'add') {
        return addKeyframe({
            expression,
            frame: operation.frame,
            value: operation.value,
        });
    }
    return removeKeyframe({ expression, frame: operation.frame });
};
const getExpressionFromJsxAttribute = (attr) => {
    if (!attr.value) {
        return b.booleanLiteral(true);
    }
    if (attr.value.type === 'StringLiteral') {
        return attr.value;
    }
    if (attr.value.type !== 'JSXExpressionContainer') {
        return null;
    }
    const { expression } = attr.value;
    if (expression.type === 'JSXEmptyExpression') {
        return null;
    }
    return expression;
};
const findJsxAttribute = (attributes, name) => {
    const attrIndex = attributes.findIndex((candidate) => {
        if (candidate.type === 'JSXSpreadAttribute') {
            return false;
        }
        if (candidate.name.type === 'JSXNamespacedName') {
            return false;
        }
        return candidate.name.name === name;
    });
    const foundAttr = attrIndex === -1 ? undefined : attributes[attrIndex];
    return {
        attrIndex,
        attr: foundAttr && foundAttr.type === 'JSXAttribute'
            ? foundAttr
            : undefined,
    };
};
const findObjectProperty = (objExpr, propertyName) => {
    const propIndex = objExpr.properties.findIndex((prop) => prop.type === 'ObjectProperty' &&
        ((prop.key.type === 'Identifier' && prop.key.name === propertyName) ||
            (prop.key.type === 'StringLiteral' && prop.key.value === propertyName)));
    return {
        propIndex,
        prop: propIndex === -1
            ? undefined
            : objExpr.properties[propIndex],
    };
};
const getObjectExpression = (attr) => {
    if (!attr.value || attr.value.type !== 'JSXExpressionContainer') {
        return null;
    }
    if (attr.value.expression.type !== 'ObjectExpression') {
        return null;
    }
    return attr.value.expression;
};
const getSequenceWritableProp = ({ attributes, key, }) => {
    const dotIndex = key.indexOf('.');
    if (dotIndex === -1) {
        const { attr: topLevelAttr } = findJsxAttribute(attributes, key);
        if (!topLevelAttr) {
            throw new Error(`Cannot update keyframes: "${key}" is not set`);
        }
        const expression = getExpressionFromJsxAttribute(topLevelAttr);
        if (!expression) {
            throw new Error(`Cannot update keyframes: "${key}" is computed`);
        }
        return {
            expression,
            setExpression: (nextExpression) => {
                topLevelAttr.value = b.jsxExpressionContainer(nextExpression);
            },
        };
    }
    const parentKey = key.slice(0, dotIndex);
    const childKey = key.slice(dotIndex + 1);
    const { attr: parentAttr } = findJsxAttribute(attributes, parentKey);
    if (!parentAttr) {
        throw new Error(`Cannot update keyframes: "${parentKey}" is not set`);
    }
    const objExpr = getObjectExpression(parentAttr);
    if (!objExpr) {
        throw new Error(`Cannot update keyframes: "${parentKey}" is computed`);
    }
    const { prop } = findObjectProperty(objExpr, childKey);
    if (!prop) {
        throw new Error(`Cannot update keyframes: "${key}" is not set`);
    }
    return {
        expression: prop.value,
        setExpression: (nextExpression) => {
            prop.value = nextExpression;
        },
    };
};
const updateSequenceKeyframesAst = ({ input, nodePath, updates, }) => {
    var _a;
    var _b;
    const ast = (0, parse_ast_1.parseAst)(input);
    const node = (0, can_update_sequence_props_1.findJsxElementAtNodePath)(ast, nodePath);
    if (!node) {
        throw new Error('Could not find a JSX element at the specified location to update keyframes');
    }
    if (!node.attributes) {
        node.attributes = [];
    }
    const oldValueStrings = [];
    for (const update of updates) {
        const prop = getSequenceWritableProp({
            attributes: node.attributes,
            key: update.key,
        });
        oldValueStrings.push(recast.print(prop.expression).code);
        prop.setExpression(applyKeyframeOperation({
            expression: prop.expression,
            operation: update.operation,
        }));
    }
    return {
        serialized: (0, parse_ast_1.serializeAst)(ast),
        oldValueStrings,
        logLine: (_b = (_a = node.loc) === null || _a === void 0 ? void 0 : _a.start.line) !== null && _b !== void 0 ? _b : 1,
    };
};
exports.updateSequenceKeyframesAst = updateSequenceKeyframesAst;
const updateSequenceKeyframes = async ({ input, nodePath, updates, prettierConfigOverride, }) => {
    const { serialized, oldValueStrings, logLine } = (0, exports.updateSequenceKeyframesAst)({
        input,
        nodePath,
        updates,
    });
    const { output, formatted } = await (0, format_file_content_1.formatFileContent)({
        input: serialized,
        prettierConfigOverride,
    });
    return { output, formatted, oldValueStrings, logLine };
};
exports.updateSequenceKeyframes = updateSequenceKeyframes;
const updateEffectKeyframesAst = ({ input, sequenceNodePath, effectIndex, updates, }) => {
    var _a, _b;
    var _c, _d, _e;
    const ast = (0, parse_ast_1.parseAst)(input);
    const jsx = (0, can_update_sequence_props_1.findJsxElementAtNodePath)(ast, sequenceNodePath);
    if (!jsx) {
        throw new Error('Could not find a JSX element at the specified location to update effect keyframes');
    }
    const attr = (0, update_effect_props_1.findEffectsAttr)((_c = jsx.attributes) !== null && _c !== void 0 ? _c : []);
    if (!attr) {
        throw new Error('Could not find effects on the target JSX element');
    }
    const found = (0, update_effect_props_1.findEffectCallExpression)({ attr, effectIndex });
    if (found.kind === 'error') {
        throw new Error(`Cannot update effect keyframe: ${found.reason}`);
    }
    const { call, callee: effectCallee } = found;
    if (call.arguments.length === 0 ||
        call.arguments[0].type !== 'ObjectExpression') {
        throw new Error('Cannot update effect keyframe: computed');
    }
    const objExpr = call.arguments[0];
    const oldValueStrings = [];
    for (const update of updates) {
        const { prop } = findObjectProperty(objExpr, update.key);
        if (!prop) {
            throw new Error(`Cannot update keyframes: "${update.key}" is not set`);
        }
        oldValueStrings.push(recast.print(prop.value).code);
        prop.value = applyKeyframeOperation({
            expression: prop.value,
            operation: update.operation,
        });
    }
    return {
        serialized: (0, parse_ast_1.serializeAst)(ast),
        oldValueStrings,
        logLine: (_e = (_d = (_a = call.loc) === null || _a === void 0 ? void 0 : _a.start.line) !== null && _d !== void 0 ? _d : (_b = jsx.loc) === null || _b === void 0 ? void 0 : _b.start.line) !== null && _e !== void 0 ? _e : 1,
        effectCallee,
    };
};
exports.updateEffectKeyframesAst = updateEffectKeyframesAst;
const updateEffectKeyframes = async ({ input, sequenceNodePath, effectIndex, updates, prettierConfigOverride, }) => {
    const { serialized, oldValueStrings, logLine, effectCallee } = (0, exports.updateEffectKeyframesAst)({
        input,
        sequenceNodePath,
        effectIndex,
        updates,
    });
    const { output, formatted } = await (0, format_file_content_1.formatFileContent)({
        input: serialized,
        prettierConfigOverride,
    });
    return { output, formatted, oldValueStrings, logLine, effectCallee };
};
exports.updateEffectKeyframes = updateEffectKeyframes;
