// @remotion/studio/dist/components/ExpandedTracksProvider.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ExpandedTracksProvider = exports.ExpandedTracksSetterContext = exports.ExpandedTracksGetterContext = void 0;
const jsx_runtime_1 = require("react/jsx-runtime");
const react_1 = require("react");
const migrate_expanded_tracks_for_subscription_key_1 = require("../helpers/migrate-expanded-tracks-for-subscription-key");
const persist_boolean_map_1 = require("../helpers/persist-boolean-map");
const timeline_node_path_key_1 = require("../helpers/timeline-node-path-key");
const SESSION_STORAGE_KEY = 'remotion.editor.expandedTracks';
const loadExpandedTracks = () => {
    return (0, persist_boolean_map_1.loadPersistedBooleanMap)(SESSION_STORAGE_KEY);
};
exports.ExpandedTracksGetterContext = (0, react_1.createContext)({
    getIsExpanded: () => {
        throw new Error('ExpandedTracksGetterContext not initialized');
    },
});
exports.ExpandedTracksSetterContext = (0, react_1.createContext)({
    toggleTrack: () => {
        throw new Error('ExpandedTracksSetterContext not initialized');
    },
    migrateExpandedTracksForSubscriptionKey: () => {
        throw new Error('ExpandedTracksSetterContext not initialized');
    },
});
const ExpandedTracksProvider = ({ children }) => {
    const [expandedTracks, setExpandedTracks] = (0, react_1.useState)(loadExpandedTracks);
    const toggleTrack = (0, react_1.useCallback)((nodePathInfo) => {
        setExpandedTracks((prev) => {
            const key = (0, timeline_node_path_key_1.timelineNodePathInfoToKey)(nodePathInfo);
            const next = (0, persist_boolean_map_1.toggleBooleanMapKey)(prev, key);
            (0, persist_boolean_map_1.persistBooleanMap)(SESSION_STORAGE_KEY, next);
            return next;
        });
    }, []);
    const migrateExpandedTracks = (0, react_1.useCallback)((oldKey, newKey) => {
        setExpandedTracks((prev) => {
            const next = (0, migrate_expanded_tracks_for_subscription_key_1.migrateExpandedTracksForSubscriptionKey)(prev, oldKey, newKey);
            if (!next) {
                return prev;
            }
            (0, persist_boolean_map_1.persistBooleanMap)(SESSION_STORAGE_KEY, next);
            return next;
        });
    }, []);
    const getterValue = (0, react_1.useMemo)(() => ({
        getIsExpanded: (nodePathInfo) => {
            var _a;
            return (_a = expandedTracks[(0, timeline_node_path_key_1.timelineNodePathInfoToKey)(nodePathInfo)]) !== null && _a !== void 0 ? _a : false;
        },
    }), [expandedTracks]);
    const setterValue = (0, react_1.useMemo)(() => ({
        toggleTrack,
        migrateExpandedTracksForSubscriptionKey: migrateExpandedTracks,
    }), [toggleTrack, migrateExpandedTracks]);
    return (jsx_runtime_1.jsx(exports.ExpandedTracksSetterContext.Provider, { value: setterValue, children: jsx_runtime_1.jsx(exports.ExpandedTracksGetterContext.Provider, { value: getterValue, children: children }) }));
};
exports.ExpandedTracksProvider = ExpandedTracksProvider;
