// @remotion/studio-server/dist/codemods/update-effect-props/update-effect-props.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.updateEffectProps = exports.updateEffectPropsAst = exports.findEffectCallExpression = exports.enumerateEffectArrayElements = exports.findEffectsAttr = void 0;
const studio_shared_1 = require("@remotion/studio-shared");
const recast = __importStar(require("recast"));
const no_react_1 = require("remotion/no-react");
const can_update_sequence_props_1 = require("../../preview-server/routes/can-update-sequence-props");
const format_file_content_1 = require("../format-file-content");
const parse_ast_1 = require("../parse-ast");
const b = recast.types.builders;
const parseValueExpression = (value) => {
    const code = `a = ${(0, studio_shared_1.stringifyDefaultProps)({ props: value, enumPaths: [] })}`;
    const ast = (0, parse_ast_1.parseAst)(code);
    const stmt = ast.program.body[0];
    if (stmt.type !== 'ExpressionStatement' ||
        stmt.expression.type !== 'AssignmentExpression') {
        throw new Error('Failed to parse effect prop value expression');
    }
    return stmt.expression.right;
};
const findEffectsAttr = (attrs) => {
    for (const attr of attrs) {
        if (attr.type !== 'JSXAttribute') {
            continue;
        }
        const a = attr;
        if (a.name.type === 'JSXIdentifier' && a.name.name === 'effects') {
            return a;
        }
    }
    return null;
};
exports.findEffectsAttr = findEffectsAttr;
const getCalleeName = (call) => {
    const { callee } = call;
    if (callee.type === 'Identifier') {
        return callee.name;
    }
    if (callee.type === 'MemberExpression' &&
        !callee.computed &&
        callee.property.type === 'Identifier') {
        return callee.property.name;
    }
    return null;
};
const enumerateEffectArrayElements = (arr) => {
    const out = [];
    for (const el of arr.elements) {
        if (el === null) {
            out.push({ kind: 'unsupported', reason: 'not-call-expression' });
            continue;
        }
        if (el.type === 'SpreadElement') {
            out.push({ kind: 'unsupported', reason: 'spread' });
            continue;
        }
        if (el.type !== 'CallExpression') {
            out.push({ kind: 'unsupported', reason: 'not-call-expression' });
            continue;
        }
        const call = el;
        const callee = getCalleeName(call);
        if (callee === null) {
            out.push({ kind: 'unsupported', reason: 'not-call-expression' });
            continue;
        }
        out.push({ kind: 'call', callee, node: call });
    }
    return out;
};
exports.enumerateEffectArrayElements = enumerateEffectArrayElements;
const findEffectCallExpression = ({ attr, effectIndex, }) => {
    if (!attr.value || attr.value.type !== 'JSXExpressionContainer') {
        return { kind: 'error', reason: 'not-call-expression' };
    }
    const expr = attr.value.expression;
    if (expr.type !== 'ArrayExpression') {
        return { kind: 'error', reason: 'not-call-expression' };
    }
    const elements = (0, exports.enumerateEffectArrayElements)(expr);
    if (effectIndex < 0 || effectIndex >= elements.length) {
        return { kind: 'error', reason: 'not-found' };
    }
    const target = elements[effectIndex];
    if (target.kind !== 'call') {
        return { kind: 'error', reason: 'not-call-expression' };
    }
    return { kind: 'ok', call: target.node, callee: target.callee };
};
exports.findEffectCallExpression = findEffectCallExpression;
const findObjectProperty = (objExpr, propertyName) => {
    const propIndex = objExpr.properties.findIndex((p) => p.type === 'ObjectProperty' &&
        ((p.key.type === 'Identifier' && p.key.name === propertyName) ||
            (p.key.type === 'StringLiteral' &&
                p.key.value === propertyName)));
    return {
        propIndex,
        prop: propIndex !== -1
            ? objExpr.properties[propIndex]
            : undefined,
    };
};
const printObjectPropertyValue = (prop) => recast
    .print(prop.value)
    .code.replace(/[\n\r\t]+/g, ' ')
    .replace(/,(\s*[}\]])/g, '$1')
    .trim();
const removeObjectProperty = ({ objExpr, propertyName, }) => {
    const { propIndex, prop } = findObjectProperty(objExpr, propertyName);
    if (!prop || propIndex === -1) {
        return null;
    }
    const valueString = printObjectPropertyValue(prop);
    objExpr.properties.splice(propIndex, 1);
    return {
        key: propertyName,
        valueString,
    };
};
const updateEffectPropsAst = ({ input, sequenceNodePath, effectIndex, update, schema, }) => {
    var _a, _b, _c, _d;
    var _e, _f, _g, _h, _j;
    const ast = (0, parse_ast_1.parseAst)(input);
    const jsx = (0, can_update_sequence_props_1.findJsxElementAtNodePath)(ast, sequenceNodePath);
    if (!jsx) {
        throw new Error('Could not find a JSX element at the specified location to update effects');
    }
    const attr = (0, exports.findEffectsAttr)((_e = jsx.attributes) !== null && _e !== void 0 ? _e : []);
    if (!attr) {
        throw new Error('Could not find effects on the target JSX element');
    }
    const found = (0, exports.findEffectCallExpression)({
        attr,
        effectIndex,
    });
    if (found.kind === 'error') {
        throw new Error(`Cannot update effect prop: ${found.reason}`);
    }
    const { call, callee: effectCallee } = found;
    const isDefault = update.defaultValue !== null &&
        JSON.stringify(update.value) === JSON.stringify(update.defaultValue);
    let objExpr;
    if (call.arguments.length === 0) {
        if (isDefault) {
            return {
                serialized: (0, parse_ast_1.serializeAst)(ast),
                oldValueString: '',
                logLine: (_g = (_f = (_a = call.loc) === null || _a === void 0 ? void 0 : _a.start.line) !== null && _f !== void 0 ? _f : (_b = jsx.loc) === null || _b === void 0 ? void 0 : _b.start.line) !== null && _g !== void 0 ? _g : 1,
                effectCallee,
                removedProps: [],
            };
        }
        objExpr = b.objectExpression([]);
        call.arguments.push(objExpr);
    }
    else if (call.arguments[0].type !== 'ObjectExpression') {
        throw new Error('Cannot update effect prop: computed');
    }
    else {
        objExpr = call.arguments[0];
    }
    const { prop } = findObjectProperty(objExpr, update.key);
    const removedProps = [];
    let oldValueString = '';
    if (prop) {
        oldValueString = recast.print(prop.value).code;
    }
    else if (update.defaultValue !== null) {
        oldValueString = JSON.stringify(update.defaultValue);
    }
    if (isDefault) {
        if (prop) {
            const idx = objExpr.properties.indexOf(prop);
            if (idx !== -1) {
                objExpr.properties.splice(idx, 1);
            }
        }
    }
    else {
        const newValueExpr = parseValueExpression(update.value);
        if (prop) {
            prop.value = newValueExpr;
        }
        else {
            objExpr.properties.push(b.objectProperty(b.identifier(update.key), newValueExpr));
        }
    }
    const fieldSchema = schema[update.key];
    if (fieldSchema && fieldSchema.type === 'enum') {
        const propsToDelete = no_react_1.NoReactInternals.findPropsToDelete({
            schema,
            key: update.key,
            value: update.value,
        });
        for (const propToDelete of propsToDelete) {
            const removed = removeObjectProperty({
                objExpr,
                propertyName: propToDelete,
            });
            if (removed) {
                removedProps.push(removed);
            }
        }
    }
    const logLine = (_j = (_h = (_c = call.loc) === null || _c === void 0 ? void 0 : _c.start.line) !== null && _h !== void 0 ? _h : (_d = jsx.loc) === null || _d === void 0 ? void 0 : _d.start.line) !== null && _j !== void 0 ? _j : 1;
    return {
        serialized: (0, parse_ast_1.serializeAst)(ast),
        oldValueString,
        logLine,
        effectCallee,
        removedProps,
    };
};
exports.updateEffectPropsAst = updateEffectPropsAst;
const updateEffectProps = async ({ input, sequenceNodePath, effectIndex, update, schema, prettierConfigOverride, }) => {
    const { serialized, oldValueString, logLine, effectCallee, removedProps } = (0, exports.updateEffectPropsAst)({
        input,
        sequenceNodePath,
        effectIndex,
        update,
        schema,
    });
    const { output, formatted } = await (0, format_file_content_1.formatFileContent)({
        input: serialized,
        prettierConfigOverride,
    });
    return {
        output,
        oldValueString,
        formatted,
        logLine,
        effectCallee,
        removedProps,
    };
};
exports.updateEffectProps = updateEffectProps;
