// @remotion/studio-server/dist/preview-server/routes/apply-codemod.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.applyCodemodHandler = void 0;
const node_fs_1 = require("node:fs");
const node_path_1 = __importDefault(require("node:path"));
const renderer_1 = require("@remotion/renderer");
const apply_codemod_to_file_1 = require("../../codemods/apply-codemod-to-file");
const simple_diff_1 = require("../../codemods/simple-diff");
const file_watcher_1 = require("../../file-watcher");
const project_info_1 = require("../project-info");
const undo_stack_1 = require("../undo-stack");
const can_update_default_props_1 = require("./can-update-default-props");
const getCodemodUndoDescription = (codemod) => {
    if (codemod.type === 'delete-composition') {
        return {
            undoMessage: `↩️  Deletion of composition "${codemod.idToDelete}"`,
            redoMessage: `↪️  Deletion of composition "${codemod.idToDelete}"`,
            entryType: codemod.type,
        };
    }
    if (codemod.type === 'rename-composition') {
        const label = `composition "${codemod.idToRename}" to "${codemod.newId}"`;
        return {
            undoMessage: `↩️  Rename of ${label}`,
            redoMessage: `↪️  Rename of ${label}`,
            entryType: codemod.type,
        };
    }
    if (codemod.type === 'duplicate-composition') {
        const label = `composition "${codemod.idToDuplicate}" to "${codemod.newId}"`;
        return {
            undoMessage: `↩️  Duplication of ${label}`,
            redoMessage: `↪️  Duplication of ${label}`,
            entryType: codemod.type,
        };
    }
    return {
        undoMessage: '↩️  Visual control change',
        redoMessage: '↪️  Visual control change',
        entryType: 'visual-control',
    };
};
const applyCodemodHandler = async ({ input: { codemod, dryRun, symbolicatedStack }, logLevel, remotionRoot, entryPoint, }) => {
    var _a;
    try {
        const time = Date.now();
        const filePath = symbolicatedStack
            ? (0, apply_codemod_to_file_1.resolveFilePathFromSymbolicatedStack)(remotionRoot, symbolicatedStack)
            : (await (0, project_info_1.getProjectInfo)(remotionRoot, entryPoint)).rootFile;
        if (!filePath) {
            throw new Error('Cannot find file for composition in project');
        }
        (0, can_update_default_props_1.checkIfTypeScriptFile)(filePath);
        const input = (0, node_fs_1.readFileSync)(filePath, 'utf-8');
        const formatted = await (0, apply_codemod_to_file_1.applyCodemodToFile)({
            filePath,
            codeMod: codemod,
        });
        const diff = (0, simple_diff_1.simpleDiff)({
            oldLines: input.split('\n'),
            newLines: formatted.split('\n'),
        });
        if (!dryRun) {
            const { entryType, undoMessage, redoMessage } = getCodemodUndoDescription(codemod);
            (0, undo_stack_1.pushToUndoStack)({
                filePath,
                oldContents: input,
                logLevel,
                remotionRoot,
                logLine: (_a = symbolicatedStack === null || symbolicatedStack === void 0 ? void 0 : symbolicatedStack.originalLineNumber) !== null && _a !== void 0 ? _a : 1,
                description: {
                    undoMessage,
                    redoMessage,
                },
                entryType,
                suppressHmrOnFileRestore: false,
            });
            (0, undo_stack_1.suppressUndoStackInvalidation)(filePath);
            (0, file_watcher_1.writeFileAndNotifyFileWatchers)(filePath, formatted, undefined);
            const end = Date.now() - time;
            const relativePath = node_path_1.default.relative(remotionRoot, filePath);
            renderer_1.RenderInternals.Log.info({ indent: false, logLevel }, renderer_1.RenderInternals.chalk.blue(`Edited ${relativePath} in ${end}ms`));
            (0, undo_stack_1.printUndoHint)(logLevel);
        }
        return {
            success: true,
            diff,
        };
    }
    catch (err) {
        return {
            success: false,
            reason: err.message,
            stack: err.stack,
        };
    }
};
exports.applyCodemodHandler = applyCodemodHandler;
