// @remotion/studio-server/dist/preview-server/routes/can-update-effect-props.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.computeEffectPropsStatusesFromFile = exports.computeEffectPropsStatusesFromContent = exports.computeEffectPropStatus = void 0;
const node_fs_1 = require("node:fs");
const parse_ast_1 = require("../../codemods/parse-ast");
const update_effect_props_1 = require("../../codemods/update-effect-props/update-effect-props");
const resolve_file_inside_project_1 = require("../../helpers/resolve-file-inside-project");
const can_update_sequence_props_1 = require("./can-update-sequence-props");
const findEffectsAttr = (jsx) => {
    for (const attr of jsx.attributes) {
        if (attr.type !== 'JSXAttribute') {
            continue;
        }
        if (attr.name.type === 'JSXIdentifier' && attr.name.name === 'effects') {
            return attr;
        }
    }
    return null;
};
const getEffectsArrayElements = (attr) => {
    if (!attr || !attr.value || attr.value.type !== 'JSXExpressionContainer') {
        return null;
    }
    const expr = attr.value.expression;
    if (expr.type !== 'ArrayExpression') {
        return null;
    }
    return (0, update_effect_props_1.enumerateEffectArrayElements)(expr);
};
const getPropsFromObjectExpression = ({ objExpr, keys, }) => {
    const out = {};
    for (const key of keys) {
        const prop = objExpr.properties.find((p) => p.type === 'ObjectProperty' &&
            ((p.key.type === 'Identifier' && p.key.name === key) ||
                (p.key.type === 'StringLiteral' &&
                    p.key.value === key)));
        if (!prop) {
            out[key] = { canUpdate: true, codeValue: undefined };
            continue;
        }
        const valueExpr = prop.value;
        if (!(0, can_update_sequence_props_1.isStaticValue)(valueExpr)) {
            out[key] = (0, can_update_sequence_props_1.getComputedStatus)(valueExpr);
            continue;
        }
        out[key] = {
            canUpdate: true,
            codeValue: (0, can_update_sequence_props_1.extractStaticValue)(valueExpr),
        };
    }
    return out;
};
const computeEffectPropStatus = ({ jsx, effectIndex, keys, }) => {
    const attr = findEffectsAttr(jsx);
    const elements = getEffectsArrayElements(attr);
    if (!elements) {
        return {
            canUpdate: false,
            effectIndex,
            reason: 'not-found',
        };
    }
    if (effectIndex < 0 || effectIndex >= elements.length) {
        return {
            canUpdate: false,
            effectIndex,
            reason: 'not-found',
        };
    }
    const target = elements[effectIndex];
    if (target.kind !== 'call') {
        return {
            canUpdate: false,
            effectIndex,
            reason: 'not-call-expression',
        };
    }
    const call = target.node;
    if (call.arguments.length === 0) {
        const emptyProps = {};
        for (const key of keys) {
            emptyProps[key] = { canUpdate: true, codeValue: undefined };
        }
        return {
            canUpdate: true,
            callee: target.callee,
            effectIndex,
            props: emptyProps,
        };
    }
    const firstArg = call.arguments[0];
    if (firstArg.type !== 'ObjectExpression') {
        return {
            canUpdate: false,
            effectIndex,
            reason: 'computed',
        };
    }
    const resolvedProps = getPropsFromObjectExpression({
        objExpr: firstArg,
        keys,
    });
    return {
        canUpdate: true,
        effectIndex,
        callee: target.callee,
        props: resolvedProps,
    };
};
exports.computeEffectPropStatus = computeEffectPropStatus;
const computeEffectPropsStatusesFromContent = ({ fileContents, sequenceNodePath, effects, keysFor, }) => {
    const ast = (0, parse_ast_1.parseAst)(fileContents);
    const jsx = (0, can_update_sequence_props_1.findJsxElementAtNodePath)(ast, sequenceNodePath);
    if (!jsx) {
        return effects.map((_effect, effectIndex) => ({
            canUpdate: false,
            effectIndex,
            reason: 'not-found',
        }));
    }
    return effects.map((effect, effectIndex) => (0, exports.computeEffectPropStatus)({
        jsx,
        effectIndex,
        keys: keysFor(effect),
    }));
};
exports.computeEffectPropsStatusesFromContent = computeEffectPropsStatusesFromContent;
const computeEffectPropsStatusesFromFile = ({ fileName, sequenceNodePath, effects, keysFor, remotionRoot, }) => {
    const { absolutePath } = (0, resolve_file_inside_project_1.resolveFileInsideProject)({
        remotionRoot,
        fileName,
        action: 'read',
    });
    const fileContents = (0, node_fs_1.readFileSync)(absolutePath, 'utf-8');
    return (0, exports.computeEffectPropsStatusesFromContent)({
        fileContents,
        sequenceNodePath,
        effects,
        keysFor,
    });
};
exports.computeEffectPropsStatusesFromFile = computeEffectPropsStatusesFromFile;
