// @remotion/studio-server/dist/preview-server/routes/can-update-sequence-props.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.computeSequencePropsStatusFromFilenameByLine = exports.computeSequencePropsStatus = exports.computeSequencePropsStatusFromContent = exports.lineColumnToNodePath = exports.findJsxElementAtNodePath = exports.getComputedStatus = exports.extractStaticValue = exports.isStaticValue = void 0;
const node_fs_1 = require("node:fs");
const renderer_1 = require("@remotion/renderer");
const recast = __importStar(require("recast"));
const parse_ast_1 = require("../../codemods/parse-ast");
const get_ast_node_path_1 = require("../../helpers/get-ast-node-path");
const resolve_file_inside_project_1 = require("../../helpers/resolve-file-inside-project");
const jsx_element_not_found_at_location_error_1 = require("../jsx-element-not-found-at-location-error");
const can_update_effect_props_1 = require("./can-update-effect-props");
const isStaticValue = (node) => {
    switch (node.type) {
        case 'NumericLiteral':
        case 'StringLiteral':
        case 'BooleanLiteral':
        case 'NullLiteral':
            return true;
        case 'UnaryExpression':
            return ((node.operator === '-' ||
                node.operator === '+') &&
                node.argument.type === 'NumericLiteral');
        case 'TSAsExpression':
            return (0, exports.isStaticValue)(node.expression);
        case 'ArrayExpression':
            return node.elements.every((el) => el !== null && el.type !== 'SpreadElement' && (0, exports.isStaticValue)(el));
        case 'ObjectExpression':
            return node.properties.every((prop) => prop.type === 'ObjectProperty' &&
                (0, exports.isStaticValue)(prop.value));
        default:
            return false;
    }
};
exports.isStaticValue = isStaticValue;
const extractStaticValue = (node) => {
    switch (node.type) {
        case 'NumericLiteral':
        case 'StringLiteral':
        case 'BooleanLiteral':
            return node.value;
        case 'NullLiteral':
            return null;
        case 'UnaryExpression': {
            const un = node;
            if (un.argument.type === 'NumericLiteral') {
                const val = un.argument.value;
                return un.operator === '-' ? -val : val;
            }
            return undefined;
        }
        case 'TSAsExpression':
            return (0, exports.extractStaticValue)(node.expression);
        case 'ArrayExpression':
            return node.elements.map((el) => {
                if (el === null || el.type === 'SpreadElement') {
                    return undefined;
                }
                return (0, exports.extractStaticValue)(el);
            });
        case 'ObjectExpression': {
            const obj = node;
            const result = {};
            for (const prop of obj.properties) {
                if (prop.type === 'ObjectProperty') {
                    const p = prop;
                    const key = p.key.type === 'Identifier'
                        ? p.key.name
                        : p.key.type === 'StringLiteral' ||
                            p.key.type === 'NumericLiteral'
                            ? String(p.key.value)
                            : undefined;
                    if (key !== undefined) {
                        result[key] = (0, exports.extractStaticValue)(p.value);
                    }
                }
            }
            return result;
        }
        default:
            return undefined;
    }
};
exports.extractStaticValue = extractStaticValue;
const getNumericValue = (node) => {
    if (node.type === 'NumericLiteral') {
        return node.value;
    }
    if (node.type === 'UnaryExpression' &&
        (node.operator === '-' || node.operator === '+') &&
        node.argument.type === 'NumericLiteral') {
        return node.operator === '-' ? -node.argument.value : node.argument.value;
    }
    if (node.type === 'TSAsExpression') {
        return getNumericValue(node.expression);
    }
    return null;
};
const getExtrapolateType = (node) => {
    if (node.type === 'StringLiteral') {
        if (node.value === 'extend' ||
            node.value === 'identity' ||
            node.value === 'clamp' ||
            node.value === 'wrap') {
            return node.value;
        }
        return null;
    }
    if (node.type === 'TSAsExpression') {
        return getExtrapolateType(node.expression);
    }
    return null;
};
const getKeyframeEasing = (node) => {
    if (node.type === 'TSAsExpression') {
        return getKeyframeEasing(node.expression);
    }
    if (node.type === 'MemberExpression' &&
        node.object.type === 'Identifier' &&
        node.object.name === 'Easing' &&
        node.property.type === 'Identifier' &&
        node.property.name === 'linear' &&
        node.computed === false) {
        return 'linear';
    }
    if (node.type !== 'CallExpression' ||
        node.callee.type !== 'MemberExpression' ||
        node.callee.object.type !== 'Identifier' ||
        node.callee.object.name !== 'Easing' ||
        node.callee.property.type !== 'Identifier' ||
        node.callee.property.name !== 'bezier' ||
        node.callee.computed) {
        return null;
    }
    if (node.arguments.length !== 4) {
        return null;
    }
    const values = node.arguments.map((arg) => {
        if (arg.type === 'ArgumentPlaceholder' ||
            arg.type === 'JSXNamespacedName' ||
            arg.type === 'SpreadElement') {
            return null;
        }
        return getNumericValue(arg);
    });
    if (values.some((v) => v === null)) {
        return null;
    }
    return values;
};
const getKeyframeEasingArray = ({ easingNode, segments, }) => {
    if (segments === 0) {
        return [];
    }
    if (easingNode.type === 'TSAsExpression') {
        return getKeyframeEasingArray({
            easingNode: easingNode.expression,
            segments,
        });
    }
    if (easingNode.type === 'ArrayExpression') {
        if (easingNode.elements.length !== segments) {
            return null;
        }
        const parsed = easingNode.elements.map((element) => {
            if (!element || element.type === 'SpreadElement') {
                return null;
            }
            return getKeyframeEasing(element);
        });
        if (parsed.some((value) => value === null)) {
            return null;
        }
        return parsed;
    }
    const easing = getKeyframeEasing(easingNode);
    if (!easing) {
        return null;
    }
    return new Array(segments).fill(easing);
};
const getInterpolationMetadata = (callExpression, keyframeCount) => {
    const segments = Math.max(0, keyframeCount - 1);
    if (callExpression.callee.type !== 'Identifier') {
        return null;
    }
    const defaultClamping = callExpression.callee.name === 'interpolateColors'
        ? {
            left: 'clamp',
            right: 'clamp',
        }
        : {
            left: 'extend',
            right: 'extend',
        };
    const defaults = {
        easing: new Array(segments).fill('linear'),
        clamping: defaultClamping,
    };
    if (callExpression.callee.name === 'interpolateColors') {
        return defaults;
    }
    if (callExpression.callee.name !== 'interpolate') {
        return null;
    }
    const optionsArg = callExpression.arguments[3];
    if (!optionsArg) {
        return defaults;
    }
    if (optionsArg.type !== 'ObjectExpression') {
        return null;
    }
    let { easing } = defaults;
    let { clamping } = defaults;
    for (const property of optionsArg.properties) {
        if (property.type !== 'ObjectProperty' || property.computed) {
            return null;
        }
        const key = property.key.type === 'Identifier'
            ? property.key.name
            : property.key.type === 'StringLiteral'
                ? property.key.value
                : null;
        if (!key) {
            return null;
        }
        const value = property.value;
        if (key === 'easing') {
            const parsedEasing = getKeyframeEasingArray({
                easingNode: value,
                segments,
            });
            if (!parsedEasing) {
                return null;
            }
            easing = parsedEasing;
            continue;
        }
        if (key === 'extrapolateLeft' || key === 'extrapolateRight') {
            const extrapolateType = getExtrapolateType(value);
            if (!extrapolateType) {
                return null;
            }
            clamping =
                key === 'extrapolateLeft'
                    ? { ...clamping, left: extrapolateType }
                    : { ...clamping, right: extrapolateType };
        }
    }
    return { easing, clamping };
};
const getInterpolationKeyframes = (node) => {
    if (node.type === 'TSAsExpression') {
        return getInterpolationKeyframes(node.expression);
    }
    if (node.type !== 'CallExpression') {
        return undefined;
    }
    const callExpression = node;
    if (callExpression.callee.type !== 'Identifier' ||
        (callExpression.callee.name !== 'interpolate' &&
            callExpression.callee.name !== 'interpolateColors')) {
        return undefined;
    }
    const inputArg = callExpression.arguments[1];
    const outputArg = callExpression.arguments[2];
    if (!inputArg ||
        !outputArg ||
        inputArg.type !== 'ArrayExpression' ||
        outputArg.type !== 'ArrayExpression') {
        return undefined;
    }
    if (inputArg.elements.length !== outputArg.elements.length) {
        return undefined;
    }
    const keyframes = [];
    for (let i = 0; i < inputArg.elements.length; i++) {
        const inputElement = inputArg.elements[i];
        const outputElement = outputArg.elements[i];
        if (!inputElement ||
            !outputElement ||
            inputElement.type === 'SpreadElement' ||
            outputElement.type === 'SpreadElement') {
            return undefined;
        }
        const frame = getNumericValue(inputElement);
        if (frame === null || !(0, exports.isStaticValue)(outputElement)) {
            return undefined;
        }
        keyframes.push({
            frame,
            value: (0, exports.extractStaticValue)(outputElement),
        });
    }
    if (keyframes.length === 0) {
        return undefined;
    }
    const metadata = getInterpolationMetadata(callExpression, keyframes.length);
    if (!metadata) {
        return undefined;
    }
    return {
        keyframes,
        easing: metadata.easing,
        clamping: metadata.clamping,
    };
};
const getComputedStatus = (node) => {
    const interpolation = getInterpolationKeyframes(node);
    if (!interpolation) {
        return { canUpdate: false, reason: 'computed' };
    }
    return {
        canUpdate: false,
        reason: 'keyframed',
        keyframes: interpolation.keyframes,
        easing: interpolation.easing,
        clamping: interpolation.clamping,
    };
};
exports.getComputedStatus = getComputedStatus;
const getPropsStatus = (jsxElement) => {
    const props = {};
    for (const attr of jsxElement.attributes) {
        if (attr.type === 'JSXSpreadAttribute') {
            continue;
        }
        if (attr.name.type === 'JSXNamespacedName') {
            continue;
        }
        const { name } = attr.name;
        if (typeof name !== 'string') {
            continue;
        }
        const { value } = attr;
        if (!value) {
            props[name] = { canUpdate: true, codeValue: true };
            continue;
        }
        if (value.type === 'StringLiteral') {
            props[name] = {
                canUpdate: true,
                codeValue: value.value,
            };
            continue;
        }
        if (value.type === 'JSXExpressionContainer') {
            const { expression } = value;
            if (expression.type === 'JSXEmptyExpression') {
                props[name] = { canUpdate: false, reason: 'computed' };
                continue;
            }
            if (!(0, exports.isStaticValue)(expression)) {
                props[name] = (0, exports.getComputedStatus)(expression);
                continue;
            }
            props[name] = {
                canUpdate: true,
                codeValue: (0, exports.extractStaticValue)(expression),
            };
            continue;
        }
        props[name] = { canUpdate: false, reason: 'computed' };
    }
    return props;
};
const getNodePathForRecastPath = (
// eslint-disable-next-line @typescript-eslint/no-explicit-any
recastPath) => {
    const segments = [];
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    let current = recastPath;
    while (current && current.parentPath) {
        segments.unshift(current.name);
        current = current.parentPath;
    }
    // Recast paths start with "root" which doesn't correspond to a real AST property
    if (segments.length > 0 && segments[0] === 'root') {
        return segments.slice(1);
    }
    return segments;
};
const findJsxElementAtNodePath = (ast, nodePath) => {
    const current = (0, get_ast_node_path_1.getAstNodePath)(ast, nodePath);
    if (!current) {
        return null;
    }
    if (recast.types.namedTypes.JSXOpeningElement.check(current.value)) {
        return current.value;
    }
    return null;
};
exports.findJsxElementAtNodePath = findJsxElementAtNodePath;
const lineColumnToNodePath = (ast, targetLine) => {
    let foundPath = null;
    recast.types.visit(ast, {
        visitJSXOpeningElement(p) {
            const { node } = p;
            if (node.loc && node.loc.start.line === targetLine) {
                foundPath = getNodePathForRecastPath(p);
                return false;
            }
            return this.traverse(p);
        },
    });
    return foundPath;
};
exports.lineColumnToNodePath = lineColumnToNodePath;
const PIXEL_VALUE_REGEX = /^-?\d+(\.\d+)?px$/;
const isSupportedTranslateValue = (value) => {
    const parts = value.split(/\s+/);
    if (parts.length === 1 || parts.length === 2) {
        return parts.every((part) => PIXEL_VALUE_REGEX.test(part));
    }
    return false;
};
const validateStyleValue = (childKey, value) => {
    if (childKey === 'translate' && typeof value === 'string') {
        return isSupportedTranslateValue(value);
    }
    return true;
};
const getNestedPropStatus = (jsxElement, parentKey, childKey) => {
    const attr = jsxElement.attributes.find((a) => a.type !== 'JSXSpreadAttribute' &&
        a.name.type !== 'JSXNamespacedName' &&
        a.name.name === parentKey);
    if (!attr || !attr.value) {
        // Parent attribute doesn't exist, nested prop can be added
        return { canUpdate: true, codeValue: undefined };
    }
    if (attr.value.type !== 'JSXExpressionContainer') {
        return { canUpdate: false, reason: 'computed' };
    }
    const { expression } = attr.value;
    if (expression.type === 'JSXEmptyExpression' ||
        expression.type !== 'ObjectExpression') {
        // Parent is not an object literal (e.g. style={myStyles})
        return { canUpdate: false, reason: 'computed' };
    }
    const objExpr = expression;
    const prop = objExpr.properties.find((p) => p.type === 'ObjectProperty' &&
        ((p.key.type === 'Identifier' && p.key.name === childKey) ||
            (p.key.type === 'StringLiteral' && p.key.value === childKey)));
    if (!prop) {
        // Property not set in the object, can be added
        return { canUpdate: true, codeValue: undefined };
    }
    const propValue = prop.value;
    if (!(0, exports.isStaticValue)(propValue)) {
        return (0, exports.getComputedStatus)(propValue);
    }
    const codeValue = (0, exports.extractStaticValue)(propValue);
    if (!validateStyleValue(childKey, codeValue)) {
        return { canUpdate: false, reason: 'computed' };
    }
    return { canUpdate: true, codeValue };
};
const computeEffectsForJsx = ({ jsxElement, effects, }) => {
    return effects.map((effect, effectIndex) => (0, can_update_effect_props_1.computeEffectPropStatus)({
        jsx: jsxElement,
        effectIndex,
        keys: effect,
    }));
};
const computeSequenceOnlyPropsRecord = ({ jsxElement, keys, }) => {
    const allProps = getPropsStatus(jsxElement);
    const filteredProps = {};
    for (const key of keys) {
        const dotIndex = key.indexOf('.');
        if (dotIndex !== -1) {
            filteredProps[key] = getNestedPropStatus(jsxElement, key.slice(0, dotIndex), key.slice(dotIndex + 1));
        }
        else if (key in allProps) {
            filteredProps[key] = allProps[key];
        }
        else {
            filteredProps[key] = { canUpdate: true, codeValue: undefined };
        }
    }
    return filteredProps;
};
const computeSequencePropsStatusFromContent = ({ fileContents, nodePath, keys, effects, }) => {
    const ast = (0, parse_ast_1.parseAst)(fileContents);
    const jsxElement = (0, exports.findJsxElementAtNodePath)(ast, nodePath);
    if (!jsxElement) {
        throw new jsx_element_not_found_at_location_error_1.JsxElementNotFoundAtLocationError();
    }
    const filteredProps = computeSequenceOnlyPropsRecord({ jsxElement, keys });
    const effectsStatuses = computeEffectsForJsx({ jsxElement, effects });
    return {
        canUpdate: true,
        props: filteredProps,
        effects: effectsStatuses,
    };
};
exports.computeSequencePropsStatusFromContent = computeSequencePropsStatusFromContent;
const computeSequencePropsStatus = ({ fileName, nodePath, keys, effects, remotionRoot, }) => {
    const { absolutePath } = (0, resolve_file_inside_project_1.resolveFileInsideProject)({
        remotionRoot,
        fileName,
        action: 'read',
    });
    const fileContents = (0, node_fs_1.readFileSync)(absolutePath, 'utf-8');
    return (0, exports.computeSequencePropsStatusFromContent)({
        fileContents,
        nodePath,
        keys,
        effects,
    });
};
exports.computeSequencePropsStatus = computeSequencePropsStatus;
const computeSequencePropsStatusFromFilenameByLine = ({ fileName, line, keys, effects, remotionRoot, logLevel, }) => {
    try {
        const { absolutePath } = (0, resolve_file_inside_project_1.resolveFileInsideProject)({
            remotionRoot,
            fileName,
            action: 'read',
        });
        const fileContents = (0, node_fs_1.readFileSync)(absolutePath, 'utf-8');
        const ast = (0, parse_ast_1.parseAst)(fileContents);
        const resolvedNodePath = (0, exports.lineColumnToNodePath)(ast, line);
        if (!resolvedNodePath) {
            return {
                status: {
                    canUpdate: false,
                    reason: 'not-found',
                },
                success: false,
            };
        }
        return {
            status: (0, exports.computeSequencePropsStatus)({
                fileName,
                nodePath: resolvedNodePath,
                keys,
                effects,
                remotionRoot,
            }),
            nodePath: {
                absolutePath,
                nodePath: resolvedNodePath,
                sequenceKeys: keys,
                effectKeys: effects,
            },
            success: true,
        };
    }
    catch (err) {
        renderer_1.RenderInternals.Log.error({ indent: false, logLevel }, err);
        return {
            status: {
                canUpdate: false,
                reason: 'error',
            },
            success: false,
        };
    }
};
exports.computeSequencePropsStatusFromFilenameByLine = computeSequencePropsStatusFromFilenameByLine;
