// @remotion/studio-server/dist/preview-server/routes/delete-effect.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.deleteEffectHandler = void 0;
const node_fs_1 = require("node:fs");
const renderer_1 = require("@remotion/renderer");
const delete_effect_1 = require("../../codemods/delete-effect");
const file_watcher_1 = require("../../file-watcher");
const resolve_file_inside_project_1 = require("../../helpers/resolve-file-inside-project");
const format_log_file_location_1 = require("../format-log-file-location");
const undo_stack_1 = require("../undo-stack");
const formatting_1 = require("./log-updates/formatting");
const log_update_1 = require("./log-updates/log-update");
const deleteEffectHandler = async ({ input: { fileName, sequenceNodePath, effectIndex }, remotionRoot, logLevel, }) => {
    try {
        renderer_1.RenderInternals.Log.trace({ indent: false, logLevel }, `[delete-effect] Received request for fileName="${fileName}" effectIndex=${effectIndex}`);
        const { absolutePath, fileRelativeToRoot } = (0, resolve_file_inside_project_1.resolveFileInsideProject)({
            remotionRoot,
            fileName,
            action: 'modify',
        });
        const fileContents = (0, node_fs_1.readFileSync)(absolutePath, 'utf-8');
        const { output, formatted, effectLabel, logLine } = await (0, delete_effect_1.deleteEffect)({
            input: fileContents,
            sequenceNodePath: sequenceNodePath.nodePath,
            effectIndex,
        });
        (0, undo_stack_1.pushToUndoStack)({
            filePath: absolutePath,
            oldContents: fileContents,
            logLevel,
            remotionRoot,
            logLine,
            description: {
                undoMessage: `↩️  Deletion of ${effectLabel}`,
                redoMessage: `↪️  Deletion of ${effectLabel}`,
            },
            entryType: 'delete-effect',
            suppressHmrOnFileRestore: false,
        });
        (0, undo_stack_1.suppressUndoStackInvalidation)(absolutePath);
        (0, file_watcher_1.writeFileAndNotifyFileWatchers)(absolutePath, output, undefined);
        const locationLabel = (0, format_log_file_location_1.formatLogFileLocation)({
            remotionRoot,
            absolutePath,
            line: logLine,
        });
        renderer_1.RenderInternals.Log.info({ indent: false, logLevel }, `${renderer_1.RenderInternals.chalk.blueBright(`${locationLabel}`)} ${(0, formatting_1.strikeThroughOrRemovedPrefix)(effectLabel)}`);
        if (!formatted) {
            (0, log_update_1.warnAboutPrettierOnce)(logLevel);
        }
        renderer_1.RenderInternals.Log.verbose({ indent: false, logLevel }, `[delete-effect] Wrote ${fileRelativeToRoot}${formatted ? ' (formatted)' : ''}`);
        (0, undo_stack_1.printUndoHint)(logLevel);
        return {
            success: true,
        };
    }
    catch (err) {
        return {
            success: false,
            reason: err.message,
            stack: err.stack,
        };
    }
};
exports.deleteEffectHandler = deleteEffectHandler;
