// @remotion/studio-server/dist/preview-server/routes/save-sequence-props.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.saveSequencePropsHandler = void 0;
const node_fs_1 = require("node:fs");
const renderer_1 = require("@remotion/renderer");
const studio_shared_1 = require("@remotion/studio-shared");
const no_react_1 = require("remotion/no-react");
const update_sequence_props_1 = require("../../codemods/update-sequence-props/update-sequence-props");
const file_watcher_1 = require("../../file-watcher");
const resolve_file_inside_project_1 = require("../../helpers/resolve-file-inside-project");
const undo_stack_1 = require("../undo-stack");
const watch_ignore_next_change_1 = require("../watch-ignore-next-change");
const can_update_sequence_props_1 = require("./can-update-sequence-props");
const format_prop_change_1 = require("./log-updates/format-prop-change");
const log_update_1 = require("./log-updates/log-update");
const save_props_mutex_1 = require("./save-props-mutex");
const saveSequencePropsHandler = ({ input: { fileName, nodePath, key, value, defaultValue, schema, clientId }, remotionRoot, logLevel, }) => (0, save_props_mutex_1.withSavePropsLock)(async () => {
    renderer_1.RenderInternals.Log.trace({ indent: false, logLevel }, `[save-sequence-props] Received request for fileName="${fileName}" key="${key}"`);
    const { absolutePath, fileRelativeToRoot } = (0, resolve_file_inside_project_1.resolveFileInsideProject)({
        remotionRoot,
        fileName,
        action: 'modify',
    });
    const fileContents = (0, node_fs_1.readFileSync)(absolutePath, 'utf-8');
    const { output, oldValueStrings, formatted, logLine, removedProps } = await (0, update_sequence_props_1.updateSequenceProps)({
        input: fileContents,
        nodePath: nodePath.nodePath,
        updates: [
            {
                key,
                value: JSON.parse(value),
                defaultValue: defaultValue !== null ? JSON.parse(defaultValue) : null,
            },
        ],
        schema: no_react_1.NoReactInternals.sequenceSchema,
    });
    const oldValueString = oldValueStrings[0];
    const newValueString = JSON.stringify(JSON.parse(value));
    const parsedDefault = defaultValue !== null ? JSON.parse(defaultValue) : null;
    const defaultValueString = parsedDefault !== null ? JSON.stringify(parsedDefault) : null;
    const normalizedOld = (0, log_update_1.normalizeQuotes)(oldValueString);
    const normalizedNew = (0, log_update_1.normalizeQuotes)(newValueString);
    const normalizedDefault = defaultValueString !== null ? (0, log_update_1.normalizeQuotes)(defaultValueString) : null;
    const undoPropChange = (0, format_prop_change_1.formatPropChange)({
        key,
        oldValueString: normalizedNew,
        newValueString: normalizedOld,
        defaultValueString: normalizedDefault,
        removedProps: [],
        addedProps: removedProps,
    });
    const redoPropChange = (0, format_prop_change_1.formatPropChange)({
        key,
        oldValueString: normalizedOld,
        newValueString: normalizedNew,
        defaultValueString: normalizedDefault,
        removedProps,
        addedProps: [],
    });
    (0, undo_stack_1.pushToUndoStack)({
        filePath: absolutePath,
        oldContents: fileContents,
        logLevel,
        remotionRoot,
        logLine,
        description: {
            undoMessage: `↩️  ${undoPropChange}`,
            redoMessage: `↪️  ${redoPropChange}`,
        },
        entryType: 'sequence-props',
        suppressHmrOnFileRestore: true,
    });
    (0, undo_stack_1.suppressUndoStackInvalidation)(absolutePath);
    (0, watch_ignore_next_change_1.suppressBundlerUpdateForFile)(absolutePath);
    (0, file_watcher_1.writeFileAndNotifyFileWatchers)(absolutePath, output, clientId);
    (0, log_update_1.logUpdate)({
        fileRelativeToRoot,
        line: logLine,
        key,
        oldValueString,
        newValueString,
        defaultValueString,
        formatted,
        logLevel,
        removedProps,
        addedProps: [],
    });
    (0, undo_stack_1.printUndoHint)(logLevel);
    const newStatus = (0, can_update_sequence_props_1.computeSequencePropsStatusFromContent)({
        fileContents: output,
        keys: (0, studio_shared_1.getAllSchemaKeys)(schema),
        nodePath: nodePath.nodePath,
        effects: [],
    });
    return {
        canUpdate: true,
        props: newStatus.props,
    };
});
exports.saveSequencePropsHandler = saveSequencePropsHandler;
